
# erase unlinked link
find . -xtype l | xargs rm

#REBASE
cat All_REBASE_Gold_Standards_20220530.txt | sed -e '/^>/!s/ //g'  -e '/^$/d'  > All_REBASE_Gold_Standards_20220530.faa
diamond makedb --in All_REBASE_Gold_Standards_20220530.faa -d All_REBASE_Gold_Standards_20220530 --threads 4

#========================================================================================================================================
# HiFi read analysis
#========================================================================================================================================
ln -s  00_original/*/*.bam   03_rename/ -n
ln -s  00_original/*/*/*.bam 03_rename/ -n

# make CCS reads from subreads
./qsub_ES.sh cpu 3 1 20 ./ccs.sh ../data/03_rename/NSBL.subreads.bam  ../data/10_HiFi/ 3 0.99
./qsub_ES.sh cpu 3 1 20 ./ccs.sh ../data/03_rename/NSOL.subreads.bam  ../data/10_HiFi/ 3 0.99
./qsub_ES.sh cpu 3 1 20 ./ccs.sh ../data/03_rename/NSBR.subreads.bam  ../data/10_HiFi/ 3 0.99
./qsub_ES.sh cpu 3 1 20 ./ccs.sh ../data/03_rename/NSGF.subreads.bam  ../data/10_HiFi/ 3 0.99
./qsub_ES.sh cpu 3 1 20 ./ccs.sh ../data/03_rename/NSGRN.subreads.bam ../data/10_HiFi/ 3 0.99
./qsub_ES.sh cpu 3 1 20 ./ccs.sh ../data/03_rename/NSGRY.subreads.bam ../data/10_HiFi/ 3 0.99
./qsub_ES.sh cpu 3 1 20 ./ccs.sh ../data/03_rename/NWPT.subreads.bam  ../data/10_HiFi/ 3 0.99
./qsub_ES.sh cpu 3 1 20 ./ccs.sh ../data/03_rename/NWRS.subreads.bam  ../data/10_HiFi/ 3 0.99
# SY (HiFi reads were generated by PacBio Sequel IIe)
# DW (HiFi reads were generated by PacBio Sequel IIe)
rename ".3-0.99.ccs" "_HiFi" *

# convet bam -> fastq
for f in ../data/10_HiFi/*_HiFi.bam; do
    ./qsub_ES.sh cpu 1 1 6   ./bam2fastq.sh ${f}
done
rename "_HiFi.fastq" ".fastq" *

# convert fastq -> fasta
for f in ../data/10_HiFi/*_HiFi.fastq; do
    #perl fastq2fasta.pl -a ${f}
    seqkit fq2fa ${f}
done

# split HiFi reads to small pieces for easy handling
for f in *.fa; do seqkit split -s 100000 ${f}; done

# convert BAM to be with kinetics (this may takes ~30 min)
for f in ../data/10_HiFi/*_HiFi.bam; do
    new=`echo ${f} | sed -e "s/_HiFi.bam/.HiFi_kinetics.bam/g"`
    if [ -e ${new} ]; then
        echo "Skip: ${new}"
        continue
    fi
    echo "${new}"
    ./qsub_ES.sh cpu 1 1 6 ~/anaconda3/envs/py38/bin/ccs-kinetics-bystrandify ${f} ${new} -j 60
done

# calculate basic statistics-----------------------------------------------------------
# seq length
conda activate py38
for f in ../data/10_HiFi/*.HiFi.fastq; do  python3 fasta_seqlen_distribution.py ${f}; done

# nonpareil
for f in ../data/10_HiFi/*.HiFi.fastq; do ./qsub_ES.sh cpu 1 1 6 ./nonpareil.sh ${f} 60; done

# taxonomic analysis---------------------------------------------------------------
# taxonomic annotation of reads (need >150 GB RAM at least)
for f in ../data/10_HiFi/*.hifi.fastq; do 
    #./qsub_ES.sh gpu 1 1 48 ./taxonomy_kaiju_search.sh ${f} NrEuk 16 ; 
     ./qsub_ES.sh gpu 1 1 48 ./taxonomy_kaiju_search.sh ${f} RefSeqNr 16 ; 
done

# make table
#mkdir ../taxonomy/NrEuk
#mv    ../taxonomy/NrEuk_* ../taxonomy/NrEuk
#python3 taxonomy_kaiju_table.py ../taxonomy/NrEuk/ 0.03  # with "Other"
#python3 taxonomy_kaiju_table.py ../taxonomy/NrEuk/ 0.0   # without "Other"

mkdir ../taxonomy/RefSeqNr
mv    ../taxonomy/RefSeqNr_* ../taxonomy/RefSeqNr
python3 taxonomy_kaiju_table.py ../taxonomy/RefSeqNr/ 0.03  # with "Other"
python3 taxonomy_kaiju_table.py ../taxonomy/RefSeqNr/ 0.0   # without "Other"

#16S rRNA sequence analysis
#./16S_search_barrnap.sh ../data/10_HiFi/SY_HiFi.fastq
#for f in ../data/41_HiFi/*.fa; do ./qsub_DDBJ.sh epyc 6 2 10 ./16S_search_barrnap.sh  ${f} 6 ; done
#nohup bash -c "for f in ../data/41_HiFi/*.fa; do ./16S_search_barrnap.sh  ${f} 6 ; done"
#for f in ../data/41_HiFi/*split/*.fa; do echo ./qsub_DDBJ.sh medium 1 2 10 ./16S_search_barrnap.sh  ${f} 1 ; done
#./qsub_DDBJ.sh medium 10 1 24 bash -c \"for f in ../data/41_HiFi/*split/*.fa\; do  ./16S_search_barrnap.sh  ${f} 10 \; done\"
qlogin -q cpu_I -l elapstim_req=8:00:00 -X
for f in ../data/10_HiFi/*.fa; do 
    ./qsub_ES.sh cpu 1 1 6  ./16S_search_barrnap.sh  ${f} 60 &
done
#./qsub_DDBJ.sh intel 10 2 10  ./16S_search_barrnap.sh ../data/10_HiFi/DW.fa 10
#./qsub_DDBJ.sh medium 10 1 24 bash -c \"for f in ../data/10_HiFi/*split/*.fa\; do  ./16S_search_barrnap.sh  ${f} 10 \; done\"

for f in ../data/10_HiFi/*.split/*.fa; do 
    ./qsub_DDBJ.sh intel 6 2 10   ./16S_search_sortmerna.sh  ${f} 6 
done

for f in ../16S/*; do
    ./taxonomy_silva_search.sh ${f} 10
done

for f in ../blast/*; do
    python3 blast_annotation_silva.py ${f}
done

#gene call------------------------------------------
for f in ../data/10_HiFi/*.fastq; do
    #./qsub_ES.sh cpu 1 1 6 ./genecall_prodigal.sh ${f} meta ; 
    ./qsub_DDBJ.sh intel 10 1 24 ./genecall_prodigal.sh ${f} meta ; 
done
#./qsub_DDBJ.sh intel 10 1 24 ./genecall_prodigal.sh  ../data/10_HiFi/DW.fa meta
mv ../gene/genecall/* ../gene/HiFi/

for f in ../gene/HiFi/*.faa; do seqkit split -s 5000000 ${f}; done

#annotation
for f in ../gene/HiFi/*.split/*.faa; do
    #./qsub_DDBJ.sh medium 4 4 24  ./fasta_similarity-search.sh -t REBASE -i ${f} -s blast -T 4
    ./qsub_ES.sh cpu 1 1 6  ./fasta_similarity-search.sh -t REBASE -i ${f} -s diamond -T 60
done

mkdir ../blast/HiFi_REBASE/work -p
mv ../blast/diamond_*part_*REBASE.tsv ../blast/HiFi_REBASE/work

#merge blast output
for f in ../blast/HiFi_REBASE/work/diamond_*part_001_REBASE.tsv; do
    filename=`basename ${f}`  #blast_NSBL.hifi_reads.part_001_REBASE.tsv
    sample=`echo ${filename} | cut -f2 -d "_" | cut -f1 -d "."`
    echo ${sample}
    cat ../blast/HiFi_REBASE/work/diamond_${sample}.cds.part_* > ../blast/HiFi_REBASE/diamond_${sample}.all.REBASE.tsv
done

#MTase annotation
for f in ../blast/HiFi_REBASE/diamond_*.all.REBASE.tsv; do
    filename=`basename ${f}`
    sample=`echo ${filename} | cut -f2 -d "_" | cut -f1 -d "."`
    echo ${sample}
    python3 MTase_summarize_enzymeTypes.py ../gene/HiFi/${sample}.cds.faa   ${f}
done
mv ../MTase/summary/ ../MTase/summary_Metagenome/


#========================================================================================================================================
# 1st assembly 
#========================================================================================================================================
#hifiasm-meta (will use 100-320 GB RAM space)
for f in ../data/10_HiFi/*.fastq; do 
    ./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ${f}  -t hifiasm_meta -x hifi -T 20 ; 
    #./qsub_ES.sh     gpu  1  1 48 ./assembly_metagenomic.sh -i ${f}  -t hifiasm_meta -x hifi -T 60 ; 
done

./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ../data/10_HiFi/DW.fastq  -t hifiasm_meta -x hifi -T 20  #this will take >50h with 20 CPU on DDBJ server, hieeee
#./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ../data/10_HiFi/SY.fastq  -t hifiasm_meta -x hifi -T 20

#will not end with <48h on ES system...
#./qsub_ES.sh gpu 1 1 48 ./assembly_metagenomic.sh -i ../data/10_HiFi/DW.fastq  -t hifiasm_meta -x hifi -T 12 ; 

rename "_HiFi" "" ../assembly/hifiasmMeta_*.fa

# move
mkdir ../assembly/1st
mv hifiasmMeta_* ../assembly/1st

# rename seqID
mkdir ../assembly/1st_rename
for f in ../assembly/1st/*.fa; do
    prefix=`basename ${f} | sed -e "s/.fa/_contig/g"`
    ./rename_seqID.sh ${f} ${prefix}
done
mv ../renameSeqID/hifiasmMeta_* ../assembly/1st_rename

# contig
cat ../assembly/1st_rename/*.fa > ../assembly/HotSpringMetaepigenomev01_Contig.fasta
seqkit seq -m 10000 ../assembly/HotSpringMetaepigenomev01_Contig.fasta > ../assembly/HotSpringMetaepigenomev01_Contig_u10k.fasta 
cp ../assembly/HotSpringMetaepigenomev01_Contig_u10k.fasta  ~/database/MyDB/
./qsub_ES.sh cpu 1 1 6 ./genecall_prodigal.sh ../assembly/HotSpringMetaepigenomev01_Contig_u10k.fasta meta
mv ../gene/genecall/../assembly/HotSpringMetaepigenomev01_Contig_u10k.pro.faa ../gene/
cp ../gene/HotSpringMetaepigenomev01_Contig_u10k.pro.faa ~/database/MyDB/

# get contig length distribution
for f in ../assembly/1st/*.fa; do  
    python3 fasta_seqlen_distribution.py ${f}; 
done

#CRISPR detection
./qsub_ES.sh cpu 1 1  6 ./CRISPR_CRISPRCasTyper.sh ../assembly/HotSpringMetaepigenomev01_Contig_u10k.fasta   meta 60
./qsub_ES.sh gpu 1 1 48 ./CRISPR_CRISPRCasTyper.sh ../assembly/HotSpringMetaepigenomev01_Contig_u10k.fasta   meta 16




#get linear contigs
mkdir ../assembly/1st_select
mkdir ../assembly/1st_linear
for f in ../assembly/1st/*.fa; do
    filename=`basename ${f} | cut -f1 -d "." `
    echo ${filename}
    cat ${f} | seqkit seq -m 500000 | seqkit grep -nrp c$ -v  > ../assembly/1st_select/1st_lineargt500k_${filename}.fasta
done

for f in ../assembly/1st_select/1st_lineargt500k_*; do
    echo ${f}
    seqkit split -s 1 ${f} -O ../assembly/1st_linear/
done

./qsub_ES.sh cpu 3 2 48 ./checkm1.sh ../assembly/1st_linear/ 128
./qsub_ES.sh cpu 1 1 6  ./checkm2.sh ../assembly/1st_linear/ 128

#get circle/complex bins   
./hifiasm_GetLargeBins.sh ../assembly/1st/hifiasmMeta_NSBL.fa  1st_NSBL
./hifiasm_GetLargeBins.sh ../assembly/1st/hifiasmMeta_NSOL.fa  1st_NSOL
./hifiasm_GetLargeBins.sh ../assembly/1st/hifiasmMeta_NSGRY.fa 1st_NSGRY
./hifiasm_GetLargeBins.sh ../assembly/1st/hifiasmMeta_NSGRN.fa 1st_NSGRN
./hifiasm_GetLargeBins.sh ../assembly/1st/hifiasmMeta_NSBR.fa  1st_NSBR
./hifiasm_GetLargeBins.sh ../assembly/1st/hifiasmMeta_NSGF.fa  1st_NSGF
./hifiasm_GetLargeBins.sh ../assembly/1st/hifiasmMeta_NWPT.fa  1st_NWPT
./hifiasm_GetLargeBins.sh ../assembly/1st/hifiasmMeta_NWRS.fa  1st_NWRS
./hifiasm_GetLargeBins.sh ../assembly/1st/hifiasmMeta_SY.fa    1st_SY
./hifiasm_GetLargeBins.sh ../assembly/1st/hifiasmMeta_DW.fa    1st_DW

#get long linear contigs
for f in ../assembly/1st/hifiasmMeta_*.fa; do
    sample=`basename ${f} | cut -f1 -d "." | cut -f2 -d "_"`
    echo ${sample}
    cat ${f} | seqkit grep -nrp c$ -v | seqkit seq -m 500000  > ../GenomeProkaryotes/ContigSepalation_WorkingSpace/1st_${sample}/1st_lineargt500k_${sample}.fasta
    seqkit split ../GenomeProkaryotes/ContigSepalation_WorkingSpace/1st_${sample}/1st_lineargt500k_${sample}.fasta -s 1 -O ../GenomeProkaryotes/ContigSepalation_WorkingSpace/1st_${sample}/1st_lineargt500k_${sample}
done

cd ../GenomeProkaryotes/
cp ../ContigSepalation_WorkingSpace/1st_*/1st_*_gt1Mb/*.fasta .

mkdir ../GenomeProkaryotes/1st_circle
cp ../GenomeProkaryotes/ContigSepalation_WorkingSpace/1st_*/1st_*_Circle/*.fasta  ../GenomeProkaryotes/1st_circle/   #circle (1th)

./qsub_ES.sh cpu 3 2 48 ./checkm1.sh ../GenomeProkaryotes/1st_circle/ 128
./qsub_ES.sh cpu 1 1  6 ./checkm2.sh ../GenomeProkaryotes/1st_circle/  64


#========================================================================================================================================
# binning for 1th complex bins
#========================================================================================================================================
# cat ../assembly/hifiasmMeta_S-OL.fa | seqkit grep -nrp "s14.ctg" > ../GenomeProkaryotes/CompBin_1st_S-OL_s14.fa
# ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/S-OL_HiFi.bam -b ../GenomeProkaryotes/CompBin_1st_S-OL_s14.fa -T 10

# ./metabat.sh  ../GenomeProkaryotes/CompBin_1st_S-OL_s14.fa ../mapping/CompBin_1st_S-OL_s14/pbmm2HiFi_CompBin_1st_S-OL_s14_S-OL_HiFi_sorted.bam
# ./metawrap.sh ../GenomeProkaryotes/CompBin_1st_S-OL_s14.fa 0 2000 NoReassembly
# ./qsub_da.sh.GPU.sh  20  ./vamb.sh ../GenomeProkaryotes/CompBin_1st_S-OL_s14.fa ../mapping/CompBin_1st_S-OL_s14/pbmm2HiFi_CompBin_1st_S-OL_s14_S-OL_HiFi_sorted.bam 20 GPU

# ./qsub_short.sh 12 ./checkm.sh ../binning/metabat_pbmm2HiFi_CompBin_1st_S-OL_s14_S-OL_HiFi_sorted/ 12


#========================================================================================================================================
#get unmapped reads for 2nd assembly  (unmapped read assembly)
#========================================================================================================================================
# #mapping and get non-mapped reads
# ./qsub_DDBJ.sh medium 8 8 24 ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/S-BL_HiFi.bam  -b ../assembly/Separate/_1st_S-BL_Circle.fasta   -T 10  -Y
# ./qsub_DDBJ.sh medium 8 8 24 ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/S-BR_HiFi.bam  -b ../assembly/Separate/_1st_S-BR_Circle.fasta   -T 10  -Y
# ./qsub_DDBJ.sh medium 8 8 24 ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/S-GF_HiFi.bam  -b ../assembly/Separate/_1st_S-GF_Circle.fasta   -T 10  -Y
# ./qsub_DDBJ.sh medium 8 8 24 ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/S-GRN_HiFi.bam -b ../assembly/Separate/_1st_S-GRN_Circle.fasta  -T 10  -Y
# ./qsub_DDBJ.sh medium 8 8 24 ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/S-GRY_HiFi.bam -b ../assembly/Separate/_1st_S-GRY_Circle.fasta  -T 10  -Y
# ./qsub_DDBJ.sh medium 8 8 24 ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/S-OL_HiFi.bam  -b ../assembly/Separate/_1st_S-OL_Circle.fasta   -T 10  -Y
# ./qsub_DDBJ.sh medium 8 8 24 ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/W-PT_HiFi.bam  -b ../assembly/Separate/_1st_W-PT_Circle.fasta   -T 10  -Y
# ./qsub_DDBJ.sh medium 8 8 24 ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/W-RS_HiFi.bam  -b ../assembly/Separate/_1st_W-RS_Circle.fasta   -T 10  -Y

# #if froget -Y option:
# for f in _*/*_sorted.bam; do  #_1st_W-RS_Circle/pbmm2HiFi__1st_W-RS_Circle_W-RS_HiFi_sorted.bam
#     echo ${f}
#     #output_mappedReadSeqID="${output_dir}/${MAPPING_TOOL}_${BASE_FILE_BASENAME}_${INPUT_FILE_BASE}_mapped.tsv"
#     #output_unmapped_fastq="${output_dir}/${MAPPING_TOOL}_${BASE_FILE_BASENAME}_${INPUT_FILE_BASE}_unmapped.fastq"
#     output_mappedReadSeqID=`echo ${f} | sed -e "s/_sorted.bam/_mapped.tsv/"`
#     output_unmapped_fastq=` echo ${f} | sed -e "s/_sorted.bam/_unmapped.fastq/"`
#     sample=`echo ${f} | cut -f3 -d "_"`
#     INPUT_FILE=../data/10_HiFi/${sample}_HiFi.fastq

#     #nohup bash -c "bedtools bamtobed -i ${f} | cut -f4 > ${output_mappedReadSeqID} " &
#     nohup seqkit grep -f ${output_mappedReadSeqID} -v ${INPUT_FILE} -o ${output_unmapped_fastq} &
# done

# #2nd assembly
# ./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ../mapping/_1st_S-BL_Circle/pbmm2HiFi__1st_S-BL_Circle_S-BL_HiFi_unmapped.fastq  -t hifiasm_meta -x hifi -T 20
# ./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ../mapping/_1st_S-BR_Circle/pbmm2HiFi__1st_S-BR_Circle_S-BR_HiFi_unmapped.fastq  -t hifiasm_meta -x hifi -T 20
# ./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ../mapping/_1st_S-GF_Circle/pbmm2HiFi__1st_S-GF_Circle_S-GF_HiFi_unmapped.fastq  -t hifiasm_meta -x hifi -T 20
# ./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ../mapping/_1st_S-GRN_Circle/pbmm2HiFi__1st_S-GRN_Circle_S-GRN_HiFi_unmapped.fastq -t hifiasm_meta -x hifi -T 20
# ./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ../mapping/_1st_S-GRY_Circle/pbmm2HiFi__1st_S-GRY_Circle_S-GRY_HiFi_unmapped.fastq -t hifiasm_meta -x hifi -T 20
# ./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ../mapping/_1st_S-OL_Circle/pbmm2HiFi__1st_S-OL_Circle_S-OL_HiFi_unmapped.fastq  -t hifiasm_meta -x hifi -T 20
# ./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ../mapping/_1st_W-PT_Circle/pbmm2HiFi__1st_W-PT_Circle_W-PT_HiFi_unmapped.fastq  -t hifiasm_meta -x hifi -T 20
# ./qsub_DDBJ.sh medium 20 24 72 ./assembly_metagenomic.sh -i ../mapping/_1st_W-RS_Circle/pbmm2HiFi__1st_W-RS_Circle_W-RS_HiFi_unmapped.fastq  -t hifiasm_meta -x hifi -T 20


#========================================================================================================================================
# get reads mapped onto complex bins for 2nd assembly (mapped-read-assembly strategy)
#========================================================================================================================================
#mapping
for f in ../GenomeProkaryotes/ContigSepalation_WorkingSpace/1st_*; do
     sample=`basename ${f} | cut -f2 -d "_"`  #1st_DW_ComplexBins -> DW
     for i in ../GenomeProkaryotes/ContigSepalation_WorkingSpace/1st_${sample}/1st_${sample}_ComplexBins/Complex*.fasta; do 
         ./qsub_ES.sh cpu 1 1 6 ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/${sample}.hifi.bam -b ${i} -T 64 -X ; 
     done
done

#When if missed -X option (not mapped reads retrieved):
#for f in ../mapping/S-OL_s*; do a=`basename ${f}`; ./qsub_epyc.sh 1 bedtools bamtofastq -i ${f}/*_sorted.bam -fq ${f}/pbmm2HiFi_${a}_S-OL_HiFi.fastq; done

#bam2fastq to retrieve mapped reads for second assemblying
qlogin -q cpu_I -l elapstim_req=8:00:00 -X
#rm ../mapping/Complex_1st_*/*.fastq
for f in ../mapping/Complex_1st_NS*; do  #dir
#for f in ../mapping/Complex_1st_NW*; do  #dir
#for f in ../mapping/Complex_1st_DW*; do  #dir
#for f in ../mapping/Complex_1st_SY*; do  #dir
    dirname=`basename ${f}`
    sample=`echo ${dirname} | cut -f3 -d "_"`  #NSOL
      bam=${f}/pbmm2HiFi_${dirname}_${sample}.hifi_sorted.bam
    fastq=${f}/pbmm2HiFi_${dirname}_${sample}.hifi.fastq
    #seqID=${f}/pbmm2HiFi_${dirname}_${sample}.aligned_seqID.tsv
    #INPUT_FILE=../data/10_HiFi/${sample}_HiFi.fastq

    #echo bedtools bamtobed -i ${bam} | cut -f4 | sort | uniq > ${seqID}
    #echo seqkit grep -f ${seqID} ${INPUT_FILE} -o ${fastq}

    # bedtools bamtobed -i pbmm2HiFi_1st_S-BL_s115_S-BL_HiFi_sorted.bam | cut -f4 | sort | uniq > seqID.tsv
    # seqkit grep -f seqID.tsv ../../data/10_HiFi/S-BL_HiFi.fastq  -o test.fastq

    echo "${bam} -> ${fastq}"
    nohup samtools fastq ${bam} --threads 10 > ${fastq} &
done


#========================================================================================================================================
# 2nd assembly (hifiasm)
#========================================================================================================================================
#for f in ../mapping/Complex_1st_*/*.fastq;  do 
for f in ../mapping/Reads_for2ndAssembly_large/*.fastq; do   #ddbj
    filename=`basename ${f} | cut -f1 -d "." `
    output_file=../assembly/2nd/hifiasm_HiFi_${filename}.fa
    echo ${output_file}
    if [ -f ${output_file} ]; then 
        echo "File already exist: ${f}"
        continue; 
    fi
    #1. normal
    #./qsub_ES.sh cpu 1 1 6 ./assembly_metagenomic.sh -i ${f} -t hifiasm -x hifi -T 64 -M 100; 

    #2. will take 7-18h
    #./qsub_ES.sh cpu 3 2 48 ./assembly_metagenomic.sh -i ${f} -t hifiasm -x hifi -T 24 -M 210; 

    #3. ROM usage >256GB
    #echo 
    #./qsub_ES.sh gpu 1 1 48 ./assembly_metagenomic.sh -i ${f} -t hifiasm -x hifi -T 16 -M 500; 

    #DDBJ
    ./qsub_DDBJ.sh medium 16 20 120 ./assembly_metagenomic.sh -i ${f} -t hifiasm -x hifi -T 16 -M 310; 
done

#some bins are moved and resumed at DDBJ server


#binning for large complex bin and get reads on each bins (i.e., sepalate reads on large complex bins )


#seqkit stat  ../assembly/hifiasmMeta_HIFI_pbmm2HiFi_1st_S-B*/*.fa
#cat   ../assembly/hifiasmMeta_HIFI_pbmm2HiFi_1st_S-B*/*.fa | grep ">"

mkdir ../assembly/2nd
mv hifiasm_HiFi_* assembly/2nd

find . -xtype l | xargs rm
for f in *.hifi; do
    filename=`basename ${f} | cut -f1 -d "." `
    ln -s ${f}/${filename}.hifi.contig.fa ${filename}.fa
done

#get long linear contigs and circler contigs
cd ../assembly/2nd/
for f in *.fa; do
    filename=`basename ${f} | cut -f1 -d "." `
    echo ${filename}

    if [ -e ../2nd_select/2nd_*_${filename}.fasta ]; then
        echo "skip"
        continue
    fi

    cat ${f} | seqkit seq -m 500000 > ../2nd_select/2nd_lineargt500k_${filename}.fasta
    seqkit grep -nrp c$ ${f}        > ../2nd_select/2nd_circle_${filename}.fasta
done

find ../2nd_select/ -type f -empty -delete

mkdir ../2nd_circle/
mkdir ../2nd_chromosome/
for f in ../2nd_select/2nd_circle_*; do
    echo ${f}
    seqkit split ${f} -s 1 -O ../2nd_circle/
done

mkdir ../2nd_linear/
for f in ../2nd_select/2nd_lineargt500k_*; do
    echo ${f}
    seqkit split -s 1 ${f} -O ../2nd_linear/
done


#========================================================================================================================================
# 3rd assembly (hifiasm)
#========================================================================================================================================
# ./hifiasm_GetLargeBins.sh ../assembly/hifiasmMeta_HIFI_pbmm2HiFi_1st_S-OL_s14_S-OL_HiFi.fa   2nd_S-OL_s14
# ./qsub_DDBJ.sh medium 12 6 24 ./mapping.sh -t pbmm2HiFi -i ../data/10_HiFi/S-OL_HiFi.bam  -b ../assembly/Separate/2nd_S-OL_s14/2nd_S-OL_s14_ComplexBins/ComplexBins/2nd_S-OL_s14_s0.fasta -T 12 -X
# ./qsub_DDBJ.sh medium 10 16 24 ./assembly_metagenomic.sh -i ../mapping/2nd_S-OL_s14_s0/pbmm2HiFi_2nd_S-OL_s14_s0_S-OL_HiFi.fastq -t hifiasm_meta -x hifi -T 10


#===================================================================================================================================================================================================
# P-MAG definition
#===================================================================================================================================================================================================
#gather P-MAG candidate contigs
mkdir ../GenomeProkaryotes/P-MAGv0.4_Circle
mkdir ../GenomeProkaryotes/P-MAGv0.4_Linear
cp ../GenomeProkaryotes/ContigSepalation_WorkingSpace/1st_*/1st_*_Circle/*.fasta        ../GenomeProkaryotes/P-MAGv0.4_Circle/   #circle (1st)
cp ../GenomeProkaryotes/ContigSepalation_WorkingSpace/1st_*/1st_lineargt500k_*/*.fasta  ../GenomeProkaryotes/P-MAGv0.4_Linear/   #long linear contig (1st)
cp ../assembly/2nd_circle/*.fasta   ../GenomeProkaryotes/P-MAGv0.4_Circle/   #circle (2nd)
cp ../assembly/2nd_linear/*.fasta   ../GenomeProkaryotes/P-MAGv0.4_Linear/   #long linear contig (2nd)

#checkM
./qsub_ES.sh cpu 3 2 48 ./checkm1.sh ../GenomeProkaryotes/P-MAGv0.4_Circle/ 128
./qsub_ES.sh cpu 3 2 48 ./checkm1.sh ../GenomeProkaryotes/P-MAGv0.4_Linear/ 128
./qsub_ES.sh cpu 1 1 6  ./checkm2.sh ../GenomeProkaryotes/P-MAGv0.4_Circle/ 128
./qsub_ES.sh cpu 1 1 6  ./checkm2.sh ../GenomeProkaryotes/P-MAGv0.4_Linear/ 128

#--------------------------
#sepalate chromosome
#--------------------------
#get chromosome
./getgenome_CheckM_result.sh ../GenomeProkaryotes/P-MAGv0.4_Circle/ ../CheckM2/checkm2_P-MAGv0.4_Circle.tsv  
./getgenome_CheckM_result.sh ../GenomeProkaryotes/P-MAGv0.4_Linear/ ../CheckM2/checkm2_P-MAGv0.4_Linear.tsv  
mkdir ../checkm_Genome/CheckM2_P-MAGv0.4_chromosome
cp  ../checkm_Genome/CheckM2_P-MAGv0.4_Circle/* ../checkm_Genome/CheckM2_P-MAGv0.4_chromosome/
cp  ../checkm_Genome/CheckM2_P-MAGv0.4_Linear/* ../checkm_Genome/CheckM2_P-MAGv0.4_chromosome/
rename "lineargt500k_hifiasm_HiFi_pbmm2HiFi_Complex_1st_" ""  ../checkm_Genome/CheckM2_P-MAGv0.4_chromosome/2nd_*
rename "lineargt500k_"                                    ""  ../checkm_Genome/CheckM2_P-MAGv0.4_chromosome/1st_*

# rename seq ID and filename
rm ../renameSeqID/ -r
mkdir ../GenomeProkaryotes/P-MAGv0.4_rename
for f in ../checkm_Genome/CheckM2_P-MAGv0.4_chromosome/*.fa; do
    # 1st_NSBL.part_001.fa          -> P-MAGv0.4_1st_NSBL_001
    # 2nd_NWPT_s20_NWPT.part_006.fa -> P-MAGv0.4_2nd_NWPT_s20_006.fa
    filename=`basename ${f}` 
    #genomeID=`basename ${f} | sed -e "s/.fa$//g"  -e "s/^/P-MAGv0.4_/g"  -e "s/.part//g"`  
    first=` echo ${filename} | cut -f1 -d "." | cut -f1-3 -d "_"`
    second=`echo ${filename} | cut -f2 -d "." | sed -e "s/part_//g"`
    genomeID=P-MAGv0.4_${first}_${second}  
    ./rename_seqID.sh ${f} ${genomeID}
    echo ../renameSeqID/${filename} ../renameSeqID/${genomeID}.fa 
    mv   ../renameSeqID/${filename} ../renameSeqID/${genomeID}.fa 
done
mv ../renameSeqID/*.fa ../GenomeProkaryotes/P-MAGv0.4_rename/
rm ../renameSeqID/*.tsv

# dereplicate (using >590 GB)
# (v0.4: This step was skipped. MAGs from differente samples were analysed independently, even if show similar genome sequence. This is aim to analysie strain-level diversity between different samples.)
# v0.6 DO this process.
./qsub_ES.sh gpu 1 2 48 ./genome_dereplication.sh ../GenomeProkaryotes/P-MAGv0.4_rename dRep_strict 16  #will take a lot of memory

#Finalize
mkdir ../GenomeProkaryotes/P-MAGv0.6_fix -p
mv ../dereplicate/P-MAGv0.4_rename_dRep_strict/dereplicated_genomes/*.fa  ../GenomeProkaryotes/P-MAGv0.6_fix/
cat   ../GenomeProkaryotes/P-MAGv0.6_fix/*.fa > ../GenomeProkaryotes/P-MAGv0.6_fix.fa

#quarity check
./qsub_ES.sh cpu 3 2 48 ./checkm1.sh ../GenomeProkaryotes/P-MAGv0.6_fix/ 128
./qsub_ES.sh cpu 1 1 6  ./checkm2.sh ../GenomeProkaryotes/P-MAGv0.6_fix/ 128

#lineage estimation
#./qsub_ES.sh gpu 1 2 48  ./gtdbtk.sh classify_wf ../GenomeProkaryotes/P-MAGv0.4_fix 16


#===================================================================================================================================================================================================
# SAG definition
#===================================================================================================================================================================================================
# File copy from bitbiome data
mkdir  ../singlecell/SAG_origin/ ../singlecell/SAG_origin_rawreads/
cp ../singlecell/analysis_bitbiome/data_repository/sag/basic/NSOL01_sc/*/*.fna.gz ../singlecell/SAG_origin/           # SAG
cp ../singlecell/analysis_bitbiome/rawdata/sag/*/*.fastq.gz                       ../singlecell/SAG_origin_rawreads/  # raw reads
seqkit stat ../singlecell/SAG_origin/*
seqkit stat ../singlecell/SAG_origin_rawreads/*

# CheckM
./qsub_ES.sh cpu 1 1 6 ./checkm1.sh ../singlecell/SAG_origin/  64
./qsub_ES.sh cpu 1 1 6 ./checkm2.sh ../singlecell/SAG_origin/  64

# Get HQ-SAGs
./getgenome_CheckM_results.sh ../singlecell/SAG_origin/ ../CheckM2/checkm2_SAG_origin.tsv
mkdir ../singlecell/P-SAGv0.4
mv ../checkm_Genome/CheckM2_SAG_origin/* ../singlecell/P-SAGv0.4
gunzip ../singlecell/P-SAGv0.4/*.gz

# get rewreads
mkdir ../singlecell/P-SAGv0.4_rawreads
for f in ../singlecell/P-SAGv0.4/*.fna; do
    SAGID=`basename ${f} | sed -e "s/.fna//g"`
    mv ../singlecell/SAG_origin_rawreads/${SAGID}_{1,2}.fastq.gz ../singlecell/P-SAGv0.4_rawreads/
done

# change file name
mkdir ../singlecell/P-SAGv0.4_renameFileName          -p
mkdir ../singlecell/P-SAGv0.4_renameFileName_rawreads -p
ls ../singlecell/P-SAGv0.4/*                     | awk '{ printf "cp %s ../singlecell/P-SAGv0.4_renameFileName/P-SAGv0.4_xxx_NSOL_%03d.fa\n",                  $0, NR }' | sh   #  NSOL01_sc-00002.fna -> P-SAGv0.4_xxx_NSOL_001.fa
ls ../singlecell/P-SAGv0.4_rawreads/*_1.fastq.gz | awk '{ printf "cp %s ../singlecell/P-SAGv0.4_renameFileName_rawreads/P-SAGv0.4_xxx_NSOL_%03d_1.fastq.gz\n", $0, NR }' | sh
ls ../singlecell/P-SAGv0.4_rawreads/*_2.fastq.gz | awk '{ printf "cp %s ../singlecell/P-SAGv0.4_renameFileName_rawreads/P-SAGv0.4_xxx_NSOL_%03d_2.fastq.gz\n", $0, NR }' | sh

# change seq ID
mkdir ../singlecell/P-SAGv0.4_renameSeqID -p
mkdir ../singlecell/P-SAGv0.4_renameSeqID_rawreads -p
for f in ../singlecell/P-SAGv0.4_renameFileName/*; do
    SAGID=`basename ${f} | sed -e "s/.fa$//g"`
    ./rename_seqID.sh ${f}  ${SAGID}
    cp ../renameSeqID/${SAGID}.fa ../singlecell/P-SAGv0.4_renameSeqID/${SAGID}.fa
done
# for f in ../singlecell/P-SAGv0.4_renameFileName_rawreads/*; do
#     SAGID=`basename ${f} | sed -e "s/.fa$//g"`
#     ./rename_seqID.sh ${f}  ${SAGID}
#     cp ../renameSeqID/${SAGID}.fastq.gz ../singlecell/P-SAGv0.4_renameSeqID_rawreads/${SAGID}.fa
# done

# move
mkdir ../singlecell/P-SAGv0.4_fix
cp  ../singlecell/P-SAGv0.4_renameSeqID/*.fa ../singlecell/P-SAGv0.4_fix/
cat ../singlecell/P-SAGv0.4_fix/*.fa       > ../singlecell/P-SAGv0.4_fix.fa

mkdir ../singlecell/P-SAGv0.4_fix_rawreads
mv ../singlecell/P-SAGv0.4_renameFileName_rawreads/*.gz ../singlecell/P-SAGv0.4_fix_rawreads/

# quality check and lineage estimation
./qsub_ES.sh cpu 1 1 6 ./checkm1.sh ../singlecell/P-SAGv0.4_fix/ 64
./qsub_ES.sh cpu 1 1 6 ./checkm2.sh ../singlecell/P-SAGv0.4_fix/ 64
./qsub_ES.sh gpu 1 1 48 ./gtdbtk.sh classify_wf ../singlecell/P-SAGv0.4_fix 16


#chloroflexus analysis --------------------------------------------------------
# get chloroflexus SAGs and make genome set
# manual
# P-SAGv0.4_xxx_NSOL_140 was not close to Chloroflexus aggregans (likely different genus), so remove from this analysis and make Chloroflexus2 dataset
# P-SAGv0.4_xxx_NSOL_216.fa showed phylogenetic placement different from others, suggesting different spieces. Remove this and make Chloroflexus3

#phylogeny (Genome)
#./qsub_ES.sh  cpu     1  1   6 ./phylophlan3.sh ../singlecell/P-SAGv0.4_Chloroflexus2/ nucl accurate low 60   #not work
./qsub_DDBJ.sh medium 16 10 120 ./phylophlan3.sh ../singlecell/P-SAGv0.4_Chloroflexus2/ nucl accurate low 60

# get chloroflexus 16S
# manual (from DFAST output)
# change SeqID

#phylogeny (16S rRNA) (fasttree)
cat ../singlecell/P-SAGv0.4_Chloroflexus2_16SrRNA/*.{fna,fa}        > ../singlecell/P-SAGv0.4_Chloroflexus2_16SrRNA.fna
cat ../singlecell/P-SAGv0.4_Chloroflexus2_16SrRNA_refine/*.{fna,fa} > ../singlecell/P-SAGv0.4_Chloroflexus2_16SrRNA_refine.fna
#./phylogenetic_tree_construction.sh  -i ../singlecell/P-SAGv0.4_Chloroflexus2_16SrRNA.fna   -S fasttree -T 40

#phylogeny (16S rRNA) (MEGA)
#manual calculation

#ANI
ls ../singlecell/P-SAGv0.4_Chloroflexus2/* > ../fastANI/P-SAGv0.4_Chloroflexus2.tsv
${HOME}/software/fastANI  --ql ../fastANI/P-SAGv0.4_Chloroflexus2.tsv --rl ../fastANI/P-SAGv0.4_Chloroflexus2.tsv -o ../fastANI/P-SAGv0.4_Chloroflexus2_fastANI.tsv -t 6

# make bai using MapSeparate bam file
samtools index ../ReadSeparation/pbmm2HiFi_P-SAGv0.4_fixmask_TOTAL/*.bam

# HiFi read mapping on the reference Chloroflexus_aggregans_DSM9485 genome
 ./qsub_ES.sh      cpu  1 1  6 ./mapping.sh -t pbmm2HiFi -b ../singlecell/P-SAGv0.4_Chloroflexus3/Chloroflexus_aggregans_DSM9485.fa  -i ../data/10_HiFi/NSOL.hifi_kinetics.bam -T 64
#./sbatch_DDBJ.sh medium 12 8 48 ./mapping.sh -t pbmm2HiFi -b ../singlecell/P-SAGv0.4_Chloroflexus3/Chloroflexus_aggregans_DSM9485.fa  -i ../data/10_HiFi/NSOL.HiFi_kinetics.bam -T 12


# SAG read mapping on the reference Chloroflexus_aggregans_DSM9485 genome
for f in ../singlecell/P-SAGv0.4_Chloroflexus3/*.fa
    GenomeID=`basename ${f}`
    ./qsub_ES.sh cpu 1 1 6 ./mapping.sh -t bowtie2 -b ../singlecell/P-SAGv0.4_Chloroflexus3/Chloroflexus_aggregans_DSM9485.fa  -1 ../singlecell/P-SAGv0.4_fix_rawreads/${GenomeID}_1.fastq.gz -2  ../singlecell/P-SAGv0.4_fix_rawreads/${GenomeID}_2.fastq.gz -T 64
done


# get short tandem repetas (STR) from reference genome
# repeat masker
# RepeatMasker -e hmmer -pa 10 ../singlecell/P-SAGv0.4_Chloroflexus2/Chloroflexus_aggregans_DSM9485.fa

# TRF
# https://github.com/Benson-Genomics-Lab/TRF
./trf409.linux64 P-SAGv0.4_Chloroflexus_aggregans/Chloroflexus_aggregans_DSM9485.fa 2 5 7 80 10 50 2000 -l 10 -ngs > TRF_Chloroflexus_aggregans_DSM9485

# get STR expansion from mapped reads
# ExpansionHunter: https://github.com/Illumina/ExpansionHunter/releases
~/workspace/software/ExpansionHunter-v5.0.0-linux_x86_64/bin/ExpansionHunter --reference  ../singlecell/P-SAGv0.4_Chloroflexus2/Chloroflexus_aggregans_DSM9485.fa --output-prefix ../singlecell/Chloroflexus_aggregans_DSM9485

# #REPrise https://github.com/hmdlab/REPrise
# git clone https://github.com/hmdlab/REPrise
# cd REPrise/
# make -j 10
# ~/workspace/software/REPrise/REPrise  -input ../singlecell/P-SAGv0.4_Chloroflexus2/Chloroflexus_aggregans_DSM9485.fa -output ../singlecell/Chloroflexus_aggregans_DSM9485 -dist 1 -minlength 2000 -pa 6


# make genome graph of chloroflexus --------------------------------------------
#pangraph
pangraph build --circular ecoli.fa.gz > ecoli_pangraph.json
pangraph export --no-duplications --output-directory ecoli_export     ecoli_pangraph.json
panta main -o examples/test/output -g examples/test/main/*.gff
panta add  -c examples/test/output -g examples/test/add/*.gff

#pggb
# https://github.com/pangenome/pggb
    #install --------------
    conda create -n pggb python=3.10
    conda activate pggb
    conda install  -c bioconda pggb==0.7.2 -c conda-forge
    pip install multiqc --user
    #----------------------
mkdir ../pangenome
conda activate pggb
cat ../singlecell/P-SAGv0.4_Chloroflexus_aggregans/* > ../singlecell/P-SAGv0.4_Chloroflexus_aggregans.fa
#bgzip -@ 3 ../singlecell/P-SAGv0.4_Chloroflexus_aggregans.fa
samtools faidx ../singlecell/P-SAGv0.4_Chloroflexus_aggregans.fa
#/home/hiraoka-s/software/pggb-0.7.2/pggb -i ../singlecell/P-SAGv0.4_Chloroflexus_aggregans.fa -o ../pangenome/P-SAGv0.4_Chloroflexus_aggregans -n 9 -t 6 -p 90 -s 5k -V 'CP001337.1:#'
pggb -i ../singlecell/P-SAGv0.4_Chloroflexus_aggregans.fa -o ../pangenome/P-SAGv0.4_Chloroflexus_aggregans -n 9 -t 6 -p 90 -s 5k 


#===================================================================================================================================================================================================
# V-MAG definition
#===================================================================================================================================================================================================
#split contig files
seqkit split -p 5 hifiasmMeta_NSBR.fa
seqkit split -p 5 hifiasmMeta_NSGRN.fa

#virsorter2
#!!!! this part can work at ES system, but not QSUB system. Use at qlogin node !!!!!!!!!!
qlogin -q cpu_I -l elapstim_req=8:00:00 -X
qlogin -q gpu_S -l elapstim_req=24:00:00 -X

#rm -r ../virsorter2/hifiasmMeta_*.part_00*
for f in ../assembly/1st_split/*split/hifiasmMeta_*.fa ; do   #1st (63)
#for f in ../assembly/2nd/hifiasm_HiFi_*.fa; do  #2nd (231)
#for f in ../assembly/1st_split/*split ; do   #parallel 1st (63)
#for f in ../assembly/2nd; do  #parallel 2nd (231)
    new_filename=`basename ${f} | sed -e "s/.fa/.out/g "`   
    if [ -e ../virsorter2/${new_filename}/final-viral-combined.fa  ]; then
        echo "Already exist. skip: ${f}"
        continue
    fi

    #./qsub_DDBJ.sh epyc 8 4 48 ./virsorter2.sh ${f} 8
    #echo     
    #./virsorter2.sh ${f} 60  # will take >10 min per file (1st)
    #./virsorter2.sh ${f} 64  # will take >10 min per file (1st)
    
    #nohup                   ./virsorter2.sh ${f} 64 & 
    #./qsub_ES.sh gpu 1 1 48 ./virsorter2.sh ${f} 16
    echo ./qsub_ES.sh cpu 1 1 3  ./virsorter2.sh ${f} 64
done

parallel -j 4 ./virsorter2.sh {} 4 ::: ../assembly/1st_split/*split/hifiasmMeta_*.fa

# cat ../virsorter2/hifiasmMeta_NSBR.part_*/final-viral-combined.fa  > ../virsorter2/hifiasmMeta_NSBR.out/final-viral-combined.fa
# cat ../virsorter2/hifiasmMeta_NSGRN.part_*/final-viral-combined.fa > ../virsorter2/hifiasmMeta_NSGRN.out/final-viral-combined.fa
# rename  hifiasmMeta_NSBR.part_   SEPARATE_hifiasmMeta_NSBR.part_   hifiasmMeta_NSBR.part_*.out
# rename  hifiasmMeta_NSGRN.part_  SEPARATE_hifiasmMeta_NSGRN.part_  hifiasmMeta_NSGRN.part_*.out

#toy
# virsorter run -w test.out -i test.fa --min-length 1500 -j 4 all
# ./qsub_ES.sh cpu 1 1 6  ./virsorter2.sh ../assembly/2nd/hifiasm_HiFi_pbmm2HiFi_Complex_1st_NWPT_s203_NWPT.fa 64
# ./qsub_ES.sh cpu 1 1 6  ./virsorter2.sh ../assembly/2nd/hifiasm_HiFi_pbmm2HiFi_Complex_1st_NWRS_s30_NWRS.fa  64
# ./qsub_ES.sh cpu 1 1 6  ./virsorter2.sh ../assembly/2nd/hifiasm_HiFi_pbmm2HiFi_Complex_1st_NSGRN_s2753_NSGRN.fa  64
# ./qsub_ES.sh cpu 1 1 1  ./virsorter2.sh ../virsorter2/test.fa  64

#rename---------------------------
cd ../virsorter2
for f in hifiasm*.out; do
    filename=`basename ${f} | sed -e "s/.fa$//g" -e "s/hifiasm_HiFi_pbmm2HiFi_Complex_//g"  -e "s/hifiasmMeta_/0th_/g" | cut -f1-3 -d "_" | sed -e "s/^/V-MAGv0.4_/g" `   #hifiasm_HiFi_pbmm2HiFi_Complex_1st_SY_s1012_SY.fa -> V-MAGsv0.4_1st_SY_s1012
    echo ${filename}
    echo cp ${f}/final-viral-combined.fa ${filename}.fa
done
rename "_1st_" "_2nd_" * ; rename "_0th_" "_1st_" *
cd ../src
mkdir ../GenomeViruses/1_V-MAGv0.4_detect/
cp ../virsorter2/*.fa  ../GenomeViruses/1_V-MAGv0.4_detect/

#change seq ID and make one big sequence file---------------------------
rm -r ../renameSeqID/
for f in ../GenomeViruses/1_V-MAGv0.4_detect/*.fa; do 
    ID=`basename ${f}| sed -e "s/.fa//g" `   #V-MAGv0.4_1st_NSBR_s118.fa -> V-MAGv0.4_1st_NSBR_s118
    echo ${ID}
    ./rename_seqID.sh ${f}  ${ID}
done

mkdir ../GenomeViruses/2_V-MAGv0.4_candidate
mv ../renameSeqID/* ../GenomeViruses/2_V-MAGv0.4_candidate/
cat ../GenomeViruses/2_V-MAGv0.4_candidate/* > ../GenomeViruses/2_V-MAGv0.4_candidate_all.fa 

# 3. sepalate by each seq ---------------------------
seqkit split -s 1 ../GenomeViruses/2_V-MAGv0.4_candidate_all.fa 
mv ../GenomeViruses/2_V-MAGv0.4_candidate_all.fa.split/* ../GenomeViruses/3_V-MAGv0.4_eachcand/

# 4. dereplicate ---------------------------
./qsub_ES.sh cpu 1 1 2 ./genome_dereplication.sh ../GenomeViruses/3_V-MAGv0.4_eachcand/ galah_virus 60
mkdir ../GenomeViruses/4_V-MAGv0.4_representative
mv  ../dereplicate/3_V-MAGv0.4_eachcand_galah_virus/*   ../GenomeViruses/4_V-MAGv0.4_representative/  
cat ../GenomeViruses/4_V-MAGv0.4_representative/*     > ../GenomeViruses/4_V-MAGv0.4_representative_all.fa

# 5. checkv and select HQ sequences---------------------------
./qsub_ES.sh   gpu    1  1 48 ./checkv.sh ../GenomeViruses/4_V-MAGv0.4_representative_all.fa  16
./qsub_DDBJ.sh intel 10 12 48 ./checkv.sh ../GenomeViruses/4_V-MAGv0.4_representative_all.fa  10

./getseq_CheckVtable.sh ../CheckV/4_V-MAGv0.4_representative_all.fa/quality_summary.tsv 
mkdir ../GenomeViruses/5_V-MAGv0.4_highquality
mv ../GenomeViruses/virus_*.fa  ../GenomeViruses/5_V-MAGv0.4_highquality/
#cat ../GenomeViruses/5_V-MAGv0.4_highquality/* > ../GenomeViruses/5_V-MAGv0.4_highquality_all.fa

# 6. Replace file name to be same with the SeqID ---------------------------
mkdir ../GenomeViruses/6_V-MAGv0.4_renameFilename/ 
for f in ../GenomeViruses/5_V-MAGv0.4_highquality/*.fa; do
    seqID=`head -n1 ${f} | cut -c2- | cut -f1 -d " "`
    echo ${seqID}
    cp ${f} ../GenomeViruses/6_V-MAGv0.4_renameFilename/${seqID}.fa
done

# 7. Add dummy "_1" in seqIDs, representing the 1st sequence on the genome. 
#    This ensure the compatible seqID format with prokaryotic/plasmid genomes ---------------------------
mkdir ../GenomeViruses/7_V-MAGv0.4_renameSeqID/ 
for f in ../GenomeViruses/6_V-MAGv0.4_renameFilename/*.fa; do
    echo ${f}
    filename=`basename ${f}`
    TMPVAR=`cat ${f} | grep -n "^>" | sed -e "s/:.*//"`  #get line nums starting with ">"
    cat ${f} | sed -e "$TMPVAR s/$\|\s/_1 /" > ../GenomeViruses/7_V-MAGv0.4_renameSeqID/${filename}  #add "_1" at the end of given lines
    #cp ${f} ../GenomeViruses/6_V-MAGv0.4_renameFilename/${seqID}.fa
done
head ../GenomeViruses/7_V-MAGv0.4_renameSeqID/V-MAGv0.4_0th_NSOL_195.fa
head ../GenomeViruses/7_V-MAGv0.4_renameSeqID/V-MAGv0.4_1st_DW_s1_16_1.fa

# copy final V-MAG set ---------------------------
mkdir ../GenomeViruses/V-MAGv0.4_fix/ 
cp  ../GenomeViruses/7_V-MAGv0.4_renameSeqID/*.fa   ../GenomeViruses/V-MAGv0.4_fix/ 
cat ../GenomeViruses/V-MAGv0.4_fix/*fa            > ../GenomeViruses/V-MAGv0.4_fix.fa

#checkv ---------------------------
./qsub_ES.sh cpu 1 1 6 ./checkv.sh ../GenomeViruses/V-MAGv0.4_fix.fa  60


#ViralRecall install #------------------------------------
cd ~/database/ViralRecall
wget -O hmm.tar.gz https://zenodo.org/record/4762520/files/hmm.tar.gz?download=1
tar -xvzf hmm.tar.gz

cd ~/software/
git clone https://github.com/faylward/viralrecall
cd viralrecall/

cd workspace/_HotSpring_metaepigenomics/src
ln -s ~/database/ViralRecall/hmm .

conda activate viralrecall
conda install pandas=1.5.3 hmmer biopython matplotlib -c bioconda


#for Megavirus using ViralRecall #------------------------------------
#python viralrecall.py -i examples/arm29B.fna -p test_outdir -t 2 -f
qlogin -q cpu_I -l elapstim_req=8:00:00 -X
#conda activate py38
conda activate viralrecall
cd ${HOME}/software/viralrecall/  # this software need to be run at the softare directory.
for f in ${HOME}//workspace/_HotSpring_metaepigenomics/assembly/1st_split/*.split/hifiasmMeta_*.fa ; do  #1st
    sample=`basename ${f}| sed -e "s/.fa//g"`
    mkdir HotSpring_${sample}
    nohup python viralrecall.py -i ${f} -p HotSpring_${sample} -t 10 -f &
done

#move output dirs
mkdir ../viralrecall/1_viralrecall-MAGv0.1 -p
mv  ${HOME}/software/viralrecall/HotSpring_*  ../viralrecall/1_viralrecall-MAGv0.1

mkdir  ../viralrecall/2_viralrecall-MAGv0.1_each -p
cp     ../viralrecall/1_viralrecall-MAGv0.1/*/*.fna      ../viralrecall/2_viralrecall-MAGv0.1_each/
awk 1  ../viralrecall/2_viralrecall-MAGv0.1_each/*.fna > ../viralrecall/2_viralrecall-MAGv0.1_each.fa
seqkit stat ../viralrecall/2_viralrecall-MAGv0.1_each.fa

#checkv 
./qsub_ES.sh cpu 1 1 6 ./checkv.sh ../viralrecall/2_viralrecall-MAGv0.1_each.fa  60


#CDS prediction------------------------------------------------------------------
for f in ../GenomeViruses/V-MAGv0.4_fix/*.fa; do  ./genecall_prodigal.sh ${f} meta all; done
mkdir ../gene/V-MAGv0.4_fix
mv ../gene/genecall/V-MAGv0.4_* ../gene/V-MAGv0.4_fix/

#RENAME
#rename V-MAGsv0.4 V-MAGv0.4 ../gene/V-MAGv0.4/V-MAG*.cds.*
#for f in ../gene/V-MAGv0.4/V-MAG*.cds.*; do echo ${f}; sed -i -e "s/V-MAGsv0.4/V-MAGv0.4/g" $f; done

#gene annotation ------------------------------------------------------------------


#taxonomic assignment ------------------------------------------------------------------

#VIRify
nextflow pull EBI-Metagenomics/emg-viral-pipeline
nextflow run  EBI-Metagenomics/emg-viral-pipeline -r v2.0.0 --fasta "../GenomeViruses/V-MAGv0.4_fix/V-MAGv0.4_0th_DW_214.fa" --cores 4 -profile local,docker


#===================================================================================================================================================================================================
# plasmid-MAG definition
#===================================================================================================================================================================================================
#get circler plasmid (including megaplasmid)

#plasmid list (manually created from CehckM2 output: circular MAGs with <10% completeness)
#plasmid-MAGv0.4_list.txt

#get
mkdir ../GenomePlasmids/1_plasmid-MAGv0.4_detected -p
cat ../SummaryInformation/plasmid-MAGv0.4_list.txt | while read line; do
  echo $line
  cp ../GenomeProkaryotes/P-MAGv0.4_Circle/${line}.fasta ../GenomePlasmids/1_plasmid-MAGv0.4_detected/
done 
wc ../SummaryInformation/plasmid-MAGv0.4_list.txt
ls ../GenomePlasmids/1_plasmid-MAGv0.4_detected/
rename ".fasta" ".fa" ../GenomePlasmids/1_plasmid-MAGv0.4_detected/*

#dereplicate
./qsub_ES.sh cpu 1 1 2 ./genome_dereplication.sh ../GenomePlasmids/1_plasmid-MAGv0.4_detected/ galah_virus 60
mv ../dereplicate/1_plasmid-MAGv0.4_detected_galah_virus ../GenomePlasmids/2_plasmid-MAGv0.4_dereplicate 

#rename
rm -r ../renameSeqID/
for f in ../GenomePlasmids/2_plasmid-MAGv0.4_dereplicate/*; do 
    #1st_NSGRY_Circle.part_052.fa
    #2nd_circle_hifiasm_HiFi_pbmm2HiFi_Complex_1st_NWPT_s1_NWPT.part_004.fa
    filename=`basename ${f}| sed -e "s/.fa$//g"`
    #ID=`basename ${f}| sed -e "s/.fa$//g"  -e "s/^/plasmid-MAGv0.4_/g"  -e "s/circle_hifiasm_HiFi_pbmm2HiFi_Complex_1st_//"  -e "s/_Circle//"  -e "s/.part//"`   #1st_DW_Circle.part_019.fa -> plasmid-MAGv0.4_1st_DW_019
    first=` echo ${filename} |cut -f1 -d "." | sed -e "s/circle_hifiasm_HiFi_pbmm2HiFi_Complex_1st_//"  -e "s/_Circle//" | cut -f1-3 -d "_"`
    second=`echo ${filename} |cut -f2 -d "." | sed -e "s/part_//"` 
    genomeID=plasmid-MAGv0.4_${first}_${second}
    echo ${genomeID}
    ./rename_seqID.sh ${f}  ${genomeID}
    mv ../renameSeqID/${filename}.fa ../renameSeqID/${genomeID}.fa
done

rm -r ../GenomePlasmids/plasmid-MAGv0.4_fix
mv   ../renameSeqID ../GenomePlasmids/plasmid-MAGv0.4_fix
rm   ../GenomePlasmids/plasmid-MAGv0.4_fix/*.tsv
cat  ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa > ../GenomePlasmids/plasmid-MAGv0.4_fix.fa 

#genecall (and 16S rRNA prediction) for plasmid----------------
for f in ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa; do
    ./genecall_prodigal.sh  ${f} meta all
    #./16S_search_barrnap.sh ${f}
done
mkdir ../gene/plasmid-MAGv0.4_fix
mv ../gene/genecall/plasmid-MAGv0.4_* ../gene/plasmid-MAGv0.4_fix


#===================================================================================================================================================================================================
# General MAG analysis 
#===================================================================================================================================================================================================
# mask low complexity region using komplexity-----------------------------
conda activate py38
mkdir ../GenomeProkaryotes/P-MAGv0.6_fixmask
mkdir ../singlecell/P-SAGv0.4_fixmask/
mkdir ../GenomeViruses/V-MAGv0.4_fixmask/
mkdir ../GenomePlasmids/plasmid-MAGv0.4_fixmask/

for f in ../GenomeProkaryotes/P-MAGv0.6_fix/*.fa; do
#for f in ../singlecell/P-SAGv0.4_fix/*.fa;    do
for f in ../GenomeViruses/V-MAGv0.4_fix/*.fa; do
#for f in ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa; do
    new_filename=`echo ${f} | sed -e "s/_fix/_fixmask/g" -e "s/.fa$/.mask.fa/g "`
    echo "${f} -> ${new_filename}"
    cat ${f} |  kz --mask --fasta --threshold 0.65 | fold > ${new_filename}
done

cat ../GenomeProkaryotes/P-MAGv0.6_fixmask/*.fa    > ../GenomeProkaryotes/P-MAGv0.6_fixmask.fa
cat ../singlecell/P-SAGv0.4_fixmask/*.fa           > ../singlecell/P-SAGv0.4_fixmask.fa
cat ../GenomeViruses/V-MAGv0.4_fixmask/*.fa        > ../GenomeViruses/V-MAGv0.4_fixmask.fa
cat ../GenomePlasmids/plasmid-MAGv0.4_fixmask/*.fa > ../GenomePlasmids/plasmid-MAGv0.4_fixmask.fa

# # Gather prokaryotic genomes -----------------------------
# mkdir ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fix/ -p
# cp    ../GenomeProkaryotes/P-MAGv0.4_fix/*            ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fix/
# cp    ../singlecell/P-SAGv0.4_fix/*                   ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fix/

# mkdir ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fixmask/ -p
# cp    ../GenomeProkaryotes/P-MAGv0.4_fixmask/*            ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fixmask/
# cp    ../singlecell/P-SAGv0.4_fixmask/*                   ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fixmask/

#merge
# cat   ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fix/*     > ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fix.fa
# cat   ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fixmask/* > ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fixmask.fa

# # Gather all genomes (naive) -----------------------------
# mkdir ../GenomeAll/AllGenomev0.4_fix/ -p
# cp    ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fix/*.fa     ../GenomeAll/AllGenomev0.6_fix/
# cp    ../GenomeViruses/V-MAGv0._fix/*.fa                   ../GenomeAll/AllGenomev0.6_fix/
# cp    ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa           ../GenomeAll/AllGenomev0.6_fix/

# mkdir ../GenomeAll/AllGenomev0.4_fixmask/ -p
# cp    ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fixmask/*.fa ../GenomeAll/AllGenomev0.6_fixmask/
# cp    ../GenomeViruses/V-MAGv0.5_fixmask/*.fa              ../GenomeAll/AllGenomev0.6_fixmask/
# cp    ../GenomePlasmids/plasmid-MAGv0.4_fixmask/*.fa       ../GenomeAll/AllGenomev0.6_fixmask/

#merge
# cat   ../GenomeAll/AllGenomev0.4_fix/*.fa >                ../GenomeAll/AllGenomev0.6_fix.fa
# cat   ../GenomeAll/AllGenomev0.4_fixmask/*.fa >            ../GenomeAll/AllGenomev0.6_fixmask.fa

# # dereplicate prokaryotic genomes from NSOL site (both P-MAGs and P-SAGs) and make new AllGenome set -----------------------------
# mkdir ../GenomeProkaryotes/siteNSBL_P-MAGandSAGv0.6_beforeDereprication/ -p
# cp ../singlecell/P-SAGv0.4_fix/P-SAGv0.4_*_NSOL_*.fa        ../GenomeProkaryotes/siteNSBL_P-MAGandSAGv0.6_beforeDereprication/ 
# cp ../GenomeProkaryotes/P-MAGv0.6_fix/P-MAGv0.4_*NSOL_*.fa  ../GenomeProkaryotes/siteNSBL_P-MAGandSAGv0.6_beforeDereprication/ 
# ./qsub_ES.sh gpu 1 2 48 ./genome_dereplication.sh           ../GenomeProkaryotes/siteNSBL_P-MAGandSAGv0.6_beforeDereprication/ dRep_strict 16  # will take a lot of memory
# file move
# mkdir ../GenomeProkaryotes/siteNSBL_P-MAGandSAGv0.6_fix/
# cp ../dereplicate/siteNSBL_P-MAGandSAGv0.6_beforeDereprication_dRep_strict/dereplicated_genomes/* ../GenomeProkaryotes/siteNSBL_P-MAGandSAGv0.6_fix/

# dereplicate MAG and SAGs and make new AllGenome set -----------------------------
mkdir ../GenomeProkaryotes/P-AllMAGandSAGv0.6_beforeDereprication/ -p
cp ../singlecell/P-SAGv0.4_fix/P-SAGv0.4_*.fa        ../GenomeProkaryotes/P-AllMAGandSAGv0.6_beforeDereprication/
cp ../GenomeProkaryotes/P-MAGv0.6_fix/P-MAGv0.4_*.fa ../GenomeProkaryotes/P-AllMAGandSAGv0.6_beforeDereprication/
./qsub_ES.sh gpu 1 2 48 ./genome_dereplication.sh    ../GenomeProkaryotes/P-AllMAGandSAGv0.6_beforeDereprication/ dRep_strict 16  # will take a lot of memory

# file move
mkdir ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/
cp ../dereplicate/P-AllMAGandSAGv0.6_beforeDereprication_dRep_strict/dereplicated_genomes/* ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/

# mask
mkdir ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fixmask/
conda activate py38
for f in ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/*.fa; do
    new_filename=`echo ${f} | sed -e "s/_fix/_fixmask/g" -e "s/.fa$/.mask.fa/g "`
    echo "${f} -> ${new_filename}"
    cat ${f} |  kz --mask --fasta --threshold 0.65 | fold  > ${new_filename}
done

# Gather prokaryotic genomes (derepricated) -----------------------------
# mkdir ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/ -p
# cp    `grep -L NSOL ../GenomeProkaryotes/P-MAGv0.6_fix/*.fa`          ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/
# cp    `grep -L NSOL ../singlecell/P-SAGv0.4_fix/*.fa`                 ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/
# cp    ../GenomeProkaryotes/siteNSBL_P-MAGandSAGv0.6_fix/*.fa          ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/

# mkdir ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fixmask/ -p
# cp    `grep -L NSOL ../GenomeProkaryotes/P-MAGv0.6_fixmask/*.mask.fa` ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fixmask/
# cp    `grep -L NSOL ../singlecell/P-SAGv0.4_fixmask/*.mask.fa`        ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fixmask/
# cp    ../GenomeProkaryotes/siteNSBL_P-MAGandSAGv0.6_fixmask/*.mask.fa ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fixmask/

#merge
cat   ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/*.fa                > ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix.fa
cat   ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fixmask/*.mask.fa       > ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fixmask.fa

# Gather all genomes (derepricated) -----------------------------
mkdir ../GenomeAll/AllGenomev0.6_fix/ -p
cp    ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/*.fa          ../GenomeAll/AllGenomev0.6_fix/ -f  
cp    ../GenomeViruses/V-MAGv0.4_fix/*.fa                       ../GenomeAll/AllGenomev0.6_fix/ -f
cp    ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa                ../GenomeAll/AllGenomev0.6_fix/ -f

mkdir ../GenomeAll/AllGenomev0.6_fixmask/ -p
cp    ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fixmask/*.mask.fa ../GenomeAll/AllGenomev0.6_fixmask/
cp    ../GenomeViruses/V-MAGv0.4_fixmask/*.mask.fa              ../GenomeAll/AllGenomev0.6_fixmask/
cp    ../GenomePlasmids/plasmid-MAGv0.4_fixmask/*.mask.fa       ../GenomeAll/AllGenomev0.6_fixmask/

#merge
cat   ../GenomeAll/AllGenomev0.6_fix/*.fa           > ../GenomeAll/AllGenomev0.6_fix.fa
cat   ../GenomeAll/AllGenomev0.6_fixmask/*.mask.fa  > ../GenomeAll/AllGenomev0.6_fixmask.fa

#taxnomy
./qsub_ES.sh gpu 1 2 48 ./gtdbtk.sh classify_wf ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/ 8

#checkm
./qsub_ES.sh cpu 3 2 48 ./checkm1.sh ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/ 64
./qsub_ES.sh cpu 1 1 6  ./checkm2.sh ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/ 64

#basic statistics
#seqkit stat -T ../GenomeProkaryotes/P-MAGv0.4_fix/*.fa          > ../SummaryInformation/summary_P-MAGv0.4_stat.tsv
#seqkit stat -T ../singlecell/P-SAGv0.4_fix/*.fa                 > ../SummaryInformation/summary_P-SAGv0.4_stat.tsv
#seqkit stat -T ../GenomeViruses/V-MAGv0.4_fix/*.fa              > ../SummaryInformation/summary_V-MAGv0.4_stat.tsv
#seqkit stat -T ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa       > ../SummaryInformation/summary_plasmid-MAGv0.4_stat.tsv
#seqkit stat -T ../GenomeProkaryotes/P-AllMAGandSAGv0.5_fix/*.fa > ../SummaryInformation/summary_P-AllMAGandSAGv0.5_stat.tsv
seqkit stat -T ../GenomeAll/AllGenomev0.6_fix/*.fa | cut -f4- -d "/" | sed -e "s/.fa//g"> ../SummaryInformation/summary_AllGenomev0.6_stat.tsv

#contig GC
#seqkit fx2tab -n -g ../GenomeProkaryotes/P-MAGv0.4_fix/*.fa          > ../SummaryInformation/summary_P-MAGv0.4_gc.tsv
#seqkit fx2tab -n -g ../singlecell/P-SAGv0.4_fix/*.fa                 > ../SummaryInformation/summary_P-SAGv0.4_gc.tsv
#seqkit fx2tab -n -g ../GenomeViruses/V-MAGv0.4_fix/*.fa              > ../SummaryInformation/summary_V-MAGv0.4_gc.tsv
#seqkit fx2tab -n -g ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa       > ../SummaryInformation/summary_plasmid-MAGv0.4_gc.tsv
#seqkit fx2tab -n -g ../GenomeAll/AllGenomev0.5_fix/*.fa              > ../SummaryInformation/summary_AllGenomev0.5_gc.tsv
#seqkit fx2tab -n -g ../GenomeProkaryotes/P-AllMAGandSAGv0.5_fix/*.fa > ../SummaryInformation/summary_P-AllMAGandSAGv0.5_gc.tsv

#genome GC
python3 fasta_GCcontent.py ../GenomeAll/AllGenomev0.6_fix/*.fa | cut -f4- -d "/" | sed -e "s/.fa//g" > ../SummaryInformation/summary_AllGenomev0.6_gc.tsv

# dfast -------------------------------------------
#cp     /usr/lib64/libidn.so.11.6.18 ~/local/lib/             # for blastp at qlogin server, ES system
#ln -s ~/local/lib/libidn.so.11.6.18 ~/local/lib/libidn.so.11 # for blastp at qlogin server, ES system

qlogin -q cpu_I -l elapstim_req=8:00:00 -X
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:~/local/lib/
nohup parallel -j 10 ./dfast.sh  {}  --load 90%  ::: ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/*.fa &
mkdir ../dfast/P-AllMAGandSAGv0.6_fix
mv ../dfast/P-{MAG,SAG}*  ../dfast/P-AllMAGandSAGv0.6_fix

# antiSMASH -------------------------------------------
qlogin -q cpu_I -l elapstim_req=8:00:00 -X
nohup parallel -j 10 ./antiSMASH.sh {} 6 --load 90% ::: ../GenomeAll/AllGenomev0.6_fix/*.fa &
mkdir ../antiSMASH/AllGenomev0.6_fix
mv ../antiigs
for f in ../assembly/1st/hifiasmMeta_*.fa; do 
    ./qsub_ES.sh gpu 1 1 10 ./antiSMASH.sh ${f} 16
done

# get CDS files after renaming seqID and filename-------------------------------------------
mkdir ../gene/P-AllMAGandSAGv0.6_fix/
for f in ../dfast/P-AllMAGandSAGv0.6_fix/* ; do
    echo ${f}
    sample=`basename ${f}`
    cat ${f}/genome.gff  | sed -e "s/^>Prodigal_/>${sample}_/g" > ../gene/P-AllMAGandSAGv0.6_fix/${sample}.genome.gff
    cat ${f}/protein.faa | sed -e "s/^>Prodigal_/>${sample}_/g" > ../gene/P-AllMAGandSAGv0.6_fix/${sample}.pro.faa
    cat ${f}/cds.fna     | sed -e "s/^>Prodigal_/>${sample}_/g" > ../gene/P-AllMAGandSAGv0.6_fix/${sample}.cds.fna
    cat ${f}/rna.fna     | sed -e "s/^>Prodigal_/>${sample}_/g" > ../gene/P-AllMAGandSAGv0.6_fix/${sample}.rna.fna
done

#gather all protein seqs
mkdir -p ../gene/AllGenomev0.6_fix/
cp ../gene/P-AllMAGandSAGv0.6_fix/*.pro.faa ../gene/AllGenomev0.6_fix/
cp ../gene/V-MAGv0.4_fix/*.pro.faa          ../gene/AllGenomev0.6_fix/
cp ../gene/plasmid-MAGv0.4_fix/*.pro.faa    ../gene/AllGenomev0.6_fix/


#select 16S rRNA-------------------------------------------
# mkdir ../16SrRNA/P-AllMAGandSAGv0.4; mv ../16SrRNA/*.fna ../16SrRNA/P-AllMAGandSAGv0.4
mkdir ../16SrRNA
for f in ../gene/P-AllMAGandSAGv0.6_fix/*.rna.fna; do  #ferived from DFAST
    echo ${f}
    sample=`basename ${f} | sed -e "s/.rna.fna$//g"`
    cat ${f} | seqkit grep -nrp 16S > ../16SrRNA/16SrRNA_${sample}.fna
done

#statistics----------------
seqkit stat ../gene/{P-AllMAGandSAGv0.6,V-MAGv0.4,plasmid-MAGv0.4}_fix/*.faa > ../SummaryInformation/summary_AllGenomev0.6_CDS.tsv

seqkit stat ../16SrRNA/*.fna > ../SummaryInformation/summary_P-AllMAGandSAGv0.6_16SrRNA.tsv

# HiFi read mapping to each genomes for modification detection ----------------------------------------------------------
for f in ../singlecell/P-SAGv0.4_fixmask/*.fa;   do  
#for f in ../GenomeAll/AllGenomev0.6_fixmask/*.fa; do  
    ./mapping.sh -t pbmm2HiFi -b ${f} -T 6 -I
done

for f in ../singlecell/P-SAGv0.4_fixmask/*.fa;   do  
#for f in ../GenomeAll/AllGenomev0.6_fixmask/*.fa; do
    sample=`  basename ${f} | cut -f3 -d "_" `
    ./qsub_ES.sh cpu 1 1 6  ./mapping.sh -t pbmm2HiFi -b ${f} -i ../data/10_HiFi/${sample}.hifi_kinetics.bam -T 64
done

mkdir ../mapping/Each_AllGenomev0.6
mv    ../mapping/{P-MAG,P-SAG,V-MAG,plasmid-MAG}*.mask/*_sorted.bam ../mapping/Each_AllGenomev0.6
rm -r ../mapping/{P-MAG,P-SAG,V-MAG,plasmid-MAG}*.mask

# HiFi read mapping to the total genome file, for coverage estimation and modification detection ----------------------------------------------------------
#./mapping.sh -t pbmm2HiFi -b ../GenomeProkaryotes/P-MAGv0.4_fix.fa              -T 6 -I
 ./mapping.sh -t pbmm2HiFi -b ../singlecell/P-SAGv0.4_fix.fa                     -T 6 -I
#./mapping.sh -t pbmm2HiFi -b ../GenomeViruses/V-MAGv0.4_fix.fa                  -T 6 -I
#./mapping.sh -t pbmm2HiFi -b ../GenomePlasmids/plasmid-MAGv0.4_fix.fa           -T 6 -I
#./mapping.sh -t pbmm2HiFi -b ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fix.fa     -T 6 -I
#./mapping.sh -t pbmm2HiFi -b ../GenomeAll/AllGenomev0.4_fix.fa                  -T 6 -I
 ./mapping.sh -t pbmm2HiFi -b ../GenomeAll/AllGenomev0.6_fix.fa                  -T 6 -I

#./mapping.sh -t pbmm2HiFi -b ../GenomeProkaryotes/P-MAGv0.4_fixmask.fa          -T 6 -I
 ./mapping.sh -t pbmm2HiFi -b ../singlecell/P-SAGv0.4_fixmask.fa                 -T 6 -I
#./mapping.sh -t pbmm2HiFi -b ../GenomeViruses/V-MAGv0.4_fixmask.fa              -T 6 -I
#./mapping.sh -t pbmm2HiFi -b ../GenomePlasmids/plasmid-MAGv0.4_fixmask.fa       -T 6 -I
#./mapping.sh -t pbmm2HiFi -b ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fixmask.fa -T 6 -I
#./mapping.sh -t pbmm2HiFi -b ../GenomeAll/AllGenomev0.4_fixmask.fa              -T 6 -I
 ./mapping.sh -t pbmm2HiFi -b ../GenomeAll/AllGenomev0.6_fixmask.fa              -T 6 -I

for f in ../data/10_HiFi/*.hifi_kinetics.bam; do
#    ./qsub_ES.sh cpu 3 1 24  ./mapping.sh -t pbmm2HiFi -b ../singlecell/P-SAGv0.4_fix.fa                     -i ${f} -T 64  #for coverage calculation
#     ./qsub_ES.sh cpu 3 1 24  ./mapping.sh -t pbmm2HiFi -b ../singlecell/P-SAGv0.4_fixmask.fa                 -i ${f} -T 64  #for epigenomic analysis
#    ./qsub_ES.sh cpu 1 1  6  ./mapping.sh -t pbmm2HiFi -b ../GenomeViruses/V-MAGv0.4_fixmask.fa              -i ${f} -T 64
#    ./qsub_ES.sh cpu 1 1  6  ./mapping.sh -t pbmm2HiFi -b ../GenomePlasmids/plasmid-MAGv0.4_fixmask.fa       -i ${f} -T 64
#    ./qsub_ES.sh cpu 3 1 24  ./mapping.sh -t pbmm2HiFi -b ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fix.fa     -i ${f} -T 64
#    ./qsub_ES.sh cpu 3 1 24  ./mapping.sh -t pbmm2HiFi -b ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fixmask.fa -i ${f} -T 64
    ./qsub_ES.sh cpu 3 1 24  ./mapping.sh -t pbmm2HiFi -b ../GenomeAll/AllGenomev0.6_fix.fa                  -i ${f} -T 64  #for coverage calculation
#    ./qsub_ES.sh cpu 3 1 24  ./mapping.sh -t pbmm2HiFi -b ../GenomeAll/AllGenomev0.6_fixmask.fa              -i ${f} -T 64  #for epigenomic analysis
done
# ./qsub_ES.sh cpu 3 1 48   ./mapping.sh -t pbmm2HiFi -b ../GenomeProkaryotes/P-MAGv0.4_fix.fa      -i ../data/10_HiFi/NSBR.hifi_kinetics.bam  -T 120


# Calculate total mapped read number for calculation of the total read-mapped ratio ----------------------------------------------------------
# This will take a bit long time
# rm ../coverage/totalMappedReads_P-MAGv0.4_fix.tsv
# rm ../coverage/totalMappedReads_P-SAGv0.4_fix.tsv
# rm ../coverage/totalMappedReads_V-MAGv0.4_fix.tsv
# rm ../coverage/totalMappedReads_plasmid-MAGv0.4_fix.tsv
# rm ../coverage/totalMappedReads_AllGenomev0.4_fix.tsv
# function get_totalMappedReads() {
#     for f in ../mapping/${1}/*hifi_kinetics_sorted.bam; do
#         a=`samtools view -c -F 0x4 -F 0x800 ${f} --threads 16`
#         echo -e "${f}\t${a}"
#         echo -e "${f}\t${a}"  >> ../coverage/totalMappedReads_${1}.tsv
#     done
# }
# get_totalMappedReads P-MAGv0.4_fix
# get_totalMappedReads P-SAGv0.4_fix
# get_totalMappedReads V-MAGv0.4_fix
# get_totalMappedReads plasmid-MAGv0.4_fix
# get_totalMappedReads AllGenomev0.4_fix  # This will tale a lot of time


# coverage calculation of each MAG using CoverM ----------------------------------------------------------
qlogin -q cpu_I -l elapstim_req=8:00:00 -X
cd ~/workspace/_HotSpring_metaepigenomics/src
conda activate py38
mkdir -p ../coverage/coverM/
for f in ../mapping/AllGenomev0.6_fix/pbmm2HiFi_*.hifi_kinetics_sorted.bam;   do  GenomeDir=../GenomeAll/AllGenomev0.6_fix/
for f in ../mapping/P-SAGv0.4_fix/pbmm2HiFi_*.hifi_kinetics_sorted.bam;       do  GenomeDir=../singlecell/P-SAGv0.4_fix/
    filename=`basename ${f} | sed -e "s/pbmm2HiFi_//g" -e "s/.hifi_kinetics_sorted.bam//g" `  #pbmm2HiFi_V-MAGv0.4_fix_NSBL.hifi_kinetics_sorted.bam -> V-MAGv0.4_fix_NSBL
    echo "${filename}"
    nohup coverm genome --bam-files ${f} --genome-fasta-directory ${GenomeDir} -x fa -m mean --min-covered-fraction 0 --contig-end-exclusion 0 -t 6 | sort > ../coverage/coverM/${filename}.tsv &
done
paste ../coverage/coverM/AllGenomev0.6_fix_* | grep -v ^\\[ | datamash transpose | sort | uniq | datamash transpose > ../coverage/MAGcoverage_AllGenomev0.6_fix.tsv
paste ../coverage/coverM/P-SAGv0.4_fix_*     | grep -v ^\\[ | datamash transpose | sort | uniq | datamash transpose > ../coverage/MAGcoverage_P-SAGv0.4_fix.tsv

# get coverages of each MAG against corresponding original samples
# 
rm ../coverage/SelectCorrespondingSample_AllGenomev0.6_fix.tsv; for f in ../coverage/coverM/AllGenomev0.6_fix_*.tsv; do
rm ../coverage/SelectCorrespondingSample_P-SAGv0.4_fix.tsv;     for f in ../coverage/coverM/P-SAGv0.4_fix_*.tsv;     do
    #echo ${f}  #P-MAGv0.4_fix_NSBR.tsv
    sample=`  basename ${f} | cut -f3 -d "_" | sed -e "s/.tsv//g"`
    dataset=` basename ${f} | cut -f1 -d "_"`
    Output=../coverage/SelectCorrespondingSample_${dataset}_fix.tsv
    cat ${f} | grep -v "^\[" | grep _${sample}_ | grep -v "^Genome"  >> ${Output}
    echo ${Output}
done

# coverage calculation of each MAG using mosdepth  ----------------------------------------------------------
# # mosdepth will output the same result to "samtools contig"
# ~/workspace/software/mosdepth -t 10 ../test  ../mapping/P-MAGv0.4_fix/pbmm2HiFi_P-MAGv0.4_fix_DW.hifi_sorted.bam
# for f in ../GenomeProkaryotes/P-MAGv0.4_fix/*.fa; do    #P-MAGv0.4_1st_NSBL_Circle.part_005.fa
# #for f in ../GenomeViruses/V-MAGv0.4_fix/*.fa; do     ##V-MAGsv0.4_0th_NSBL_14.fa
# #for f in ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa; do    
#     echo ${f}
#     MAGname=` basename ${f} | sed -e "s/.fa$//g" -e "s/.mask//g"` 
#     sample=`  basename ${f} | cut -f3 -d "_"` 
#     prefix=`  basename ${f} | cut -f1 -d "_"` 
#     # ~/workspace/software/mosdepth -t 10 ../mosdepth/${MAGname}  ../mapping/${MAGname}/pbmm2HiFi_${MAGname}_${sample}.hifi_kinetics_sorted.bam
# done

# for f in ../mapping/P-MAGv0.4_fix/pbmm2HiFi_*.hifi_sorted.bam;      do  FileGenome=../GenomeProkaryotes/P-MAGv0.4_fix
# #for f in ../mapping/V-MAGv0.4_fix/pbmm2HiFi_*.hifi_sorted.bam;      do  FileGenome=../GenomeViruses/V-MAGv0.4_fix
# #for f in ../mapping/plasmid-MAGv0.4_fix/pbmm2HiFi_*.hifi_sorted.bam; do  FileGenome=../GenomePlasmids/plasmid-MAGv0.4_fix
#     filename=`basename ${f} | sed -e "s/pbmm2HiFi_//g" -e "s/.hifi_sorted.bam//g" `  #pbmm2HiFi_V-MAGv0.4_fix_NSBL.hifi_sorted.bam -> V-MAGv0.4_fix_NSBL
#     echo "${filename}"    
# done

# # get contig-mappedreads table -------------------------------------------
# # will take ~15 min
# mkdir ../coverage/genome-mappedreads/ -p
# for f in ../mapping/P-MAGv0.4_fix/pbmm2HiFi_*.hifi_sorted.bam; do
# echo ${f}
# filename=`basename ${f}    | sed -e "s/pbmm2HiFi_//g" -e "s/.hifi_sorted.bam//g" `  #pbmm2HiFi_V-MAGv0.4_fix_NSBL.hifi_sorted.bam -> V-MAGv0.4_fix_NSBL
#     sample=`  echo ${filename} | cut -f3 -d "_"`
#     nohup bedtools bamtobed -i ${f}  | cut -f1,4 | grep ${sample} > ../coverage/genome-mappedreads/${filename}.tsv &
#     #P-MAGv0.4_1st_DW_003_1  231012  231260  m64483e_230119_123937/117639374/ccs     1       -
# done 

# # get reads that aligned to each MAG ----------------------------------------------------------
# mkdir ../coverage/alignedReadsToEachMAG/ -p
# mkdir ../coverage/EachMAG-mappedreads/ -p
# for f in ../GenomeProkaryotes/P-MAGv0.4_fix/*.fa; do    #P-MAGv0.4_1st_NSBL_Circle.part_005.fa
#     echo ${f}
#     MAGname=` basename ${f} | sed -e "s/.fa$//g" -e "s/.mask//g"` 
#     sample=`  basename ${f} | cut -f3 -d "_"` 
#     prefix=`  basename ${f} | cut -f1 -d "_"`  
#     cat ../coverage/genome-mappedreads/P-MAGv0.4_fix_${sample}.tsv | grep ${MAGname} | cut -f2 > ../coverage/EachMAG-mappedreads/${MAGname}.tsv
#     #seqkit grep -f ../coverage/EachMAG-mappedreads/${MAGname}.tsv   ../data/10_HiFi/${sample}.hifi_kinetics.bam  >  ../coverage/alignedReadsToEachMAG/${filename}.bam
#     #handyTool extractID my.bam -l listOfIdentifiers > output
#     samtools view -@ 10 ../data/10_HiFi/${sample}.hifi_kinetics.bam | fgrep -w -f ../coverage/EachMAG-mappedreads/${MAGname}.tsv  >  ../coverage/alignedReadsToEachMAG/${filename}.bam
# done

# #separate total BAM files in each MAG -------------------------------------------
# for f in ../GenomeProkaryotes/P-MAGv0.4_fix/*; do
#     echo ${f}
#     MAGname=` basename ${f} | sed -e "s/.fa$//g" -e "s/.mask//g"` 
#     sample=`  basename ${f} | cut -f3 -d "_"` 
#     prefix=`  basename ${f} | cut -f1 -d "_"`  
#     samtools view -@ 10 -b ../mapping/P-MAGv0.4_fix/pbmm2HiFi_P-MAGv0.4_fix_${sample}.hifi_sorted.bam -r ${MAGname}_1 > ../coverage/alignedReadsToEachMAG/${MAGname}.bam
# done

# sepalate total BAM files in each MAG using bamtools -------------------------------------------
# qlogin -q cpu_I -l elapstim_req=8:00:00 -X
# for f in ../mapping/P-MAGv0.4_fix/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do  # will take ?h
# #for f in ../mapping/plasmid-MAGv0.4_fix/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do  #will take ~1h
#     echo ${f}
#     filename=`basename ${f}    | sed -e "s/pbmm2HiFi_//g" -e "s/.hifi_kinetics_sorted.bam//g" `  #pbmm2HiFi_V-MAGv0.4_fix_NSBL.hifi_sorted.bam -> V-MAGv0.4_fix_NSBL
#     sample=`  echo ${filename} | cut -f3 -d "_"`
#     nohup bamtools split -in ${f} -reference -stub ../coverage/bamSeparateInEachMAG/${filename} &
# done
# # nohup bamtools split -in ../mapping/P-MAGv0.4_fix/pbmm2HiFi_NSBR.hifi_kinetics_sorted.bam -reference -stub ../coverage/bamSeparateInEachMAG/${filename} &


#===================================================================================================================================================================================================
# Mapped read separation by MAG 
#===================================================================================================================================================================================================
#-------------------------------------------
# Method 1.
# separate total BAM files in each MAG using elprep + bamtools  
# #-------------------------------------------
# qlogin -q cpu_I -l elapstim_req=8:00:00 -X
# cd workspace/_HotSpring_metaepigenomics/src/
# conda activate py38
# #for f in           ../mapping/P-MAGv0.4_fix/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do  # will take ?h
# #for f in           ../mapping/V-MAGv0.4_fix/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do  # will take ?h
# #for f in     ../mapping/plasmid-MAGv0.4_fix/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do  # will take ?h
# #for f in       ../mapping/P-MAGv0.4_fixmask/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do  # will take ?h
# for f in        ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do  # will take ?h
# #for f in       ../mapping/V-MAGv0.4_fixmask/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do  # will take 30 min
# #for f in ../mapping/plasmid-MAGv0.4_fixmask/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do  # will take 30 min
#     echo ${f}
#     filename=`basename ${f}    | sed -e "s/pbmm2HiFi_//g" -e "s/.hifi_kinetics_sorted.bam//g" `  #pbmm2HiFi_V-MAGv0.4_fix_NSBL.hifi_sorted.bam -> V-MAGv0.4_fix_NSBL
#     sample=`  echo ${filename} | cut -f3 -d "_"`
#     nohup elprep split ${f} ../coverage/bamSeparateInEachMAG_elprep/ --output-prefix "${filename}" --output-type bam --nr-of-threads 10 --single-end &
# done

# qlogin -q cpu_I -l elapstim_req=8:00:00 -X
# mkdir ../coverage//bamSeparateInEachMAG_elprep2bamtools -p
# #for f in ../coverage/bamSeparateInEachMAG_elprep/P-MAGv0.4_fix*; do   #P-MAGv0.4_fix_NSGF-group00025.bam
# #for f in ../coverage/bamSeparateInEachMAG_elprep/V-MAGv0.4_fix*; do   
# #for f in ../coverage/bamSeparateInEachMAG_elprep/plasmid-MAGv0.4_fix_*; do  
# #for f in ../coverage/bamSeparateInEachMAG_elprep/P-MAGv0.4_fixmask*; do
# for f in ../coverage/bamSeparateInEachMAG_elprep/P-SAGv0.4_fixmask*; do
# #for f in ../coverage/bamSeparateInEachMAG_elprep/V-MAGv0.4_fixmask*; do     #generally fast, but some will take >30 min
# #for f in ../coverage/bamSeparateInEachMAG_elprep/plasmid-MAGv0.4_fixmask_*; do  
#     echo ${f}
#     filename=`basename ${f}`
#     if [ -e ../coverage/bamSeparateInEachMAG_elprep2bamtools/${filename}.check ]; then
#         echo "skip: ${f}"
#         continue
#     fi
#     nohup bash -c "bamtools split -in ${f} -reference -stub ../coverage/bamSeparateInEachMAG_elprep2bamtools/${filename} ; \
#                    touch ../coverage/bamSeparateInEachMAG_elprep2bamtools/${filename}.check "  &
#     #sleep 1
#     sleep 0.3
# done

# # Copy HiFi read files of each sample; Reads from the same samples were mapped to MAGs that derived from the same sample.
# # The files will be renamed related to the MAG files, like pbmm2HiFi_P-MAGv0.4_1st_DW_003_DW.hifi_kinetics_sorted.bam
# mkdir -p ../coverage/alignedReadsToEachMAG_elprep2bamtools
# #for f in ../GenomeProkaryotes/P-MAGv0.4_fix/*.fa; do
# #for f in ../GenomeViruses/V-MAGv0.4_fix/*.fa; do     # V-MAGsv0.4_0th_NSBL_14.fa V-MAGv0.4_1st_SY_s6_12_1.fa 
# #for f in ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa; do  
# #for f in ../GenomeProkaryotes/P-MAGv0.4_fixmask/*.fa; do
# for f in ../singlecell/P-SAGv0.4_fixmask/*.fa; do
# #for f in ../GenomeViruses/V-MAGv0.4_fixmask/*.fa; do  
# #for f in ../GenomePlasmids/plasmid-MAGv0.4_fixmask/*.fa; do  
#     echo ${f}
#     MAGname=` basename ${f} | sed -e "s/.fa$//g" -e "s/.mask//g"` 
#     state=`   echo ${f} | rev | cut -f2 -d "/" | rev | cut -f2 -d "_"`  #fix or fixmask
#     sample=`  basename ${f} | cut -f3 -d "_"` 
#     prefix=`  basename ${f} | cut -f1 -d "_"`  
#     # P-MAGv0.4_fix_SY.REF_P-MAGv0.4_1st_NSGRN_008_1.bam 
#     # P-MAGv0.4_fix_SY-group00012.bam.REF_P-MAGv0.4_1st_NSBR_Circle_004_1.bam  
#     # plasmid-MAGv0.4_fixmask_SY-group00001.bam.REF_plasmid-MAGv0.4_1st_DW_005_1.bam
#     # plasmid-MAGv0.4_fixmask_SY-group00063.bam.REF_plasmid-MAGv0.4_2nd_NWPT_s1695_008_1.bam
#     # 

#     #for V-MAGs
#     echo cp ../coverage/bamSeparateInEachMAG_elprep2bamtools/${prefix}_${state}_${sample}-group*.bam.REF_${MAGname}.bam    ../coverage/alignedReadsToEachMAG_elprep2bamtools/aligned${state}_${MAGname}_${sample}.bam  -n

#     #for plasmid-MAGs and P-MAGs
#     #cp ../coverage/bamSeparateInEachMAG_elprep2bamtools/${prefix}_${state}_${sample}-group*.bam.REF_${MAGname}_1.bam  ../coverage/alignedReadsToEachMAG_elprep2bamtools/aligned${state}_${MAGname}_${sample}.bam  -n
# done

#-------------------------------------------
# Method 2.
# separate total BAM files in each MAG using my original way: list up mapped SeqID and extract from original BAM.  
#-------------------------------------------
#get genome-seqID list
# #for f in ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do  # will take ?h
# for f in ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted.bam; do 
#     echo "BAM: ${f}"
#     #samtools view ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted.bam | cut -f1-9 |head
#     #m64088_211023_100220/179044496/ccs/fwd  2064    P-SAGv0.4_xxx_NSOL_001_2        105     0
#     #m64088_211023_100220/100338195/ccs/fwd  0       P-SAGv0.4_xxx_NSOL_001_2        3799    29
#     #m64088_211023_100220/100338195/ccs/rev  16      P-SAGv0.4_xxx_NSOL_001_2        3799    29
#     #m64088_211023_100220/69533915/ccs/rev   0       P-SAGv0.4_xxx_NSOL_001_2        11314   60
#     #m64088_211023_100220/69533915/ccs/fwd   16      P-SAGv0.4_xxx_NSOL_001_2        11314   60
#     # get reads that showed FLAG == 0 and 16
#     samtools view ${f} -@ 8 | cut -f1-3 > ${f}.c1-3.tsv
# done

# get seqIDs per each genome (avoid)
# mkdir ../coverage/Genome-SeqID/ -p
# while read line; do
#     genome=`echo "${line}" | cut -f3`
#     seqID=` echo "${line}" | cut -f1`
#     #echo ${genome}    
#     #check multiple-hit
#done < ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted.bam.c1-3.tsv

# #get seqs according to the list
# for f in ../singlecell/P-SAGv0.4_fixmask/*.fa; do
#     echo ${f}
#     MAGname=` basename ${f} | sed -e "s/.fa$//g" -e "s/.mask//g"` 
#     state=`   echo ${f} | rev | cut -f2 -d "/" | rev | cut -f2 -d "_"`  #fix or fixmask
#     sample=`  basename ${f} | cut -f3 -d "_"` 
#     prefix=`  basename ${f} | cut -f1 -d "_"`  
#     samtools view file.bam | fgrep -w -f IDs.txt
# done

# get seqs using samtools
# mkdir ../ReadSeparation/tmp -p
# #samtools view -h ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_random1000.bam  chr20 | samtools view -bS > ../ReadSeparation/chr20_NSOL_random1000.bam
# for f in ../singlecell/P-SAGv0.4_fix/*.fa; do  ##V-MAGsv0.4_0th_NSBL_14.fa
#     genome=`basename ${f} | sed -e "s/.fa$//g"`  
#     sample=`echo ${genome} | cut -f3 -d "_"`
#     #get seq lit
#     cat ${f} | grep "^>" | cut -c2- | cut -f1 -d " " > ../ReadSeparation/tmp/${genome}.tsv

#     #echo "samtools view -h ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_${sample}.hifi_kinetics_sorted.bam  ${genome} --threads 8 | samtools view -bS --threads 8 > ../ReadSeparation/${genome}_NSOL.bam"
#     #echo "samtools view  -o ../ReadSeparation/${genome}_NSOL.bam -O BAM -b ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_${sample}.hifi_kinetics_sorted.bam  ${genome} --threads 8"
#     echo "samtools view  -o ../ReadSeparation/${genome}_NSOL.bam -O BAM -b ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_${sample}.hifi_kinetics_sorted.bam -R  ../ReadSeparation/tmp/${genome}.tsv  --threads 8"
# done

# samtools faidx genome.fa chr3:15000-20000 > chr3_15000-20000.fasta
# samtools view -b -o ../ReadSeparation/P-SAGv0.4_xxx_NSOL_224_NSOL.bam -O BAM --threads 8 ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted.bam  P-SAGv0.4_xxx_NSOL_224_4  | samtools view -bS --threads 8 -o ../ReadSeparation/P-SAGv0.4_xxx_NSOL_224.bam
# samtools view ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted.bam P-SAGv0.4_xxx_NSOL_224_4 -b > ../ReadSeparation/out.bam
# samtools view ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted.bam  P-SAGv0.4_xxx_NSOL_001_4 | less


#-------------------------------------------
# Method 3.
# Get mapped SAM lines using original script.
#-------------------------------------------
qlogin -q cpu_I -l elapstim_req=8:00:00 -X

#make toy BAM file
# bamtools random -n 1000 -in ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSGF.hifi_kinetics_sorted.bam -out ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSGF.hifi_kinetics_sorted_random1000.bam
# samtools view -bs 0.01      ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSGF.hifi_kinetics_sorted.bam -o   ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSGF.hifi_kinetics_sorted_random0.01.bam
# bamtools count -in ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSGF.hifi_kinetics_sorted_random0.01.bam

# BAM -> SAM
#for f in ../mapping/P-SAGv0.4_fixmask/*.hifi_kinetics_sorted.bam;      do
for f in ../mapping/AllGenomev0.6_fix/*.hifi_kinetics_sorted.bam; do
#for f in ../mapping/AllGenomev0.6_fixmask/*.hifi_kinetics_sorted.bam; do
    echo ${f}
    output_sam=`echo ${f} | sed -e "s/.bam$/.sam/g"`
    samtools view -h ${f} -o ${output_sam} --threads 60
done
#samtools view -h ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSGF.hifi_kinetics_sorted_random0.01.bam -@ 10 -o ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSGF.hifi_kinetics_sorted_random0.01.sam

# Separation using my original script
conda activate py38
#for f in ../mapping/P-SAGv0.4_fixmask/*.hifi_kinetics_sorted.sam;      do
for f in ../mapping/AllGenomev0.6_fix/*.hifi_kinetics_sorted.sam; do
#for f in ../mapping/AllGenomev0.6_fixmask/*.hifi_kinetics_sorted.sam; do
    echo ${f}
    nohup python3 sam_ReadsSeparatePerGenome.py ${f} &  #will take 1-3 h for AllGenomev0.6
done
#python3 sam_ReadsSeparatePerGenome.py ../mapping/P-SAGv0.4_fixmask/pbmm2HiFi_P-SAGv0.4_fixmask_NSGF.hifi_kinetics_sorted_random0.01.sam

#extract files that both the genome and reads are derived from the same sample
mkdir ../ReadSeparation/pbmm2HiFi_P-SAGv0.4_fixmask_TOTAL ;     cd ../ReadSeparation/pbmm2HiFi_P-SAGv0.4_fixmask_TOTAL
mkdir ../ReadSeparation/pbmm2HiFi_AllGenomev0.6_fix_TOTAL     ; cd ../ReadSeparation/pbmm2HiFi_AllGenomev0.6_fix_TOTAL
mkdir ../ReadSeparation/pbmm2HiFi_AllGenomev0.6_fixmask_TOTAL ; cd ../ReadSeparation/pbmm2HiFi_AllGenomev0.6_fixmask_TOTAL

for f in ../MapSep-pbmm2HiFi_P-SAGv0.4_fixmask*.hifi_kinetics_sorted/*.sam;     do
for f in ../MapSep-pbmm2HiFi_AllGenomev0.6_fix*.hifi_kinetics_sorted/*.sam;     do
#for f in ../MapSep-pbmm2HiFi_AllGenomev0.6_fixmask*.hifi_kinetics_sorted/*.sam; do
    filename=`basename ${f}` # MapSep-pbmm2HiFi_AllGenomev0.4_fixmask_DW.hifi_kinetics_sorted_plasmid-MAGv0.4_1st_DW_009.bam
    sample_reads=` echo ${filename} | cut -f4 -d "_" | cut -f1 -d "."`
    sample_genome=`echo ${filename} | cut -f9 -d "_" `
    if [ ! ${sample_reads} == ${sample_genome} ]; then
        continue
    fi
    echo ${f} ${sample_reads} ${sample_genome}
    ln -s ${f} ${filename}
done
cd ../../src

# SAM -> BAM 
for f in ../ReadSeparation/pbmm2HiFi_P-SAGv0.4_fixmask_TOTAL/*.sam;      do
for f in ../ReadSeparation/pbmm2HiFi_AllGenomev0.6_fix_TOTAL/*.sam;     do
#for f in ../ReadSeparation/pbmm2HiFi_AllGenomev0.6_fixmask_TOTAL/*.sam; do
    echo ${f}
    output_bam=`echo ${f} | sed -e "s/.sam$/.bam/g"`
    samtools view -bS ${f} -o ${output_bam} --threads 60
done
#samtools view -h ../ReadSeparation/MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSGF.hifi_kinetics_sorted_random0.01/P-SAGv0.4_xxx_NSOL_006.bam

# remove SAMs and original dir
rm ../ReadSeparation/pbmm2HiFi_P-SAGv0.4_fixmask_TOTAL/*.sam
rm ../ReadSeparation/MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_*.hifi_kinetics_sorted     -r
rm ../mapping/AllGenomev0.6_fix/*.hifi_kinetics_sorted.sam
rm ../ReadSeparation/pbmm2HiFi_AllGenomev0.6_fix_TOTAL/*.sam
rm ../ReadSeparation/MapSep-pbmm2HiFi_AllGenomev0.6_fix_*.hifi_kinetics_sorted     -r
rm ../mapping/AllGenomev0.6_fixmask/*.hifi_kinetics_sorted.sam
rm ../ReadSeparation/pbmm2HiFi_AllGenomev0.6_fixmask_TOTAL/*.sam
rm ../ReadSeparation/MapSep-pbmm2HiFi_AllGenomev0.6_fixmask_*.hifi_kinetics_sorted -r

# #check 20250618
# cd ../mapping
# nohup samtools view AllGenomev0.5_fix/pbmm2HiFi_AllGenomev0.5_fix_NSBR.hifi_kinetics_sorted.bam -@ 20 | cut -f1-6 > pbmm2HiFi_AllGenomev0.5_fix_NSBR.hifi_kinetics_sorted.head.tsv &
# ls pbmm2HiFi_AllGenomev0.5_fix_NSBR.hifi_kinetics_sorted.head.tsv  -la

# cat pbmm2HiFi_AllGenomev0.5_fix_NSBR.hifi_kinetics_sorted.head.tsv | grep P-MAGv0.4_2nd_NSBR_ | less
# cat pbmm2HiFi_AllGenomev0.5_fix_NSBR.hifi_kinetics_sorted.head.tsv | grep P-MAGv0.4_1st_NSBR_ | less

#===================================================================================================================================================================================================
# Phylogenetic analysis
#===================================================================================================================================================================================================
#P-MAG-----------------------
#./qsub_ES.sh cpu 1 1 6 ./phylogenetic_tree_construction.sh -i ../GTDB-Tk/de_novo_wf_ALL_P-MAGv0.4_chromosome/align/gtdbtk.ar53.msa.fasta   -S fasttree -T 40
#./qsub_ES.sh cpu 1 1 6 ./phylogenetic_tree_construction.sh -i ../GTDB-Tk/de_novo_wf_ALL_P-MAGv0.4_chromosome/align/gtdbtk.bac120.msa.fasta -S fasttree -T 40

#phylophlan should be run at DDBJ server, not ES
#./phylophlan3.sh ../GenomeProkaryotes/P-MAGv0.4_toy/ nucl fast     high 16   #test
#./sbatch_DDBJ.sh medium 16 10 120 ./phylophlan3.sh ../GenomeProkaryotes/P-MAGv0.4_fix/          nucl accurate high 16   #will take >50h
#./sbatch_DDBJ.sh medium 16 10 120 ./phylophlan3.sh ../GenomeProkaryotes/P-SAGv0.4_fix/          nucl accurate high 16   #will take >50h
 ./sbatch_DDBJ.sh medium 16 10 120 ./phylophlan3.sh ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/ nucl accurate high 16   #will take >?h

#V-MAG------------------------------------------------------------------
# !!!!!!!!!!!! !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
# This will be run at the login server in DDBJ, not ES 
# This is because of R package dependencies... 
# Will take 1 h
# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
mkdir ../viptree
conda activate py39
      ${HOME}/software/ViPTreeGen-1.1.3/ViPTreeGen ../viptree/V-MAG_2Jul2022.fa      ../viptree/V-MAG_2Jul2022/ --ncpus  8
nohup ${HOME}/software/ViPTreeGen-1.1.3/ViPTreeGen ../GenomeViruses/V-MAGv0.4_fix.fa ../viptree/V-MAGv0.4_fix/  --ncpus 12 &

#./qsub_ES.sh cpu 1 1 6       bash -c \"source ${HOME}/anaconda3/etc/profile.d/conda.sh\; conda activate py39\; ${HOME}/software/ViPTreeGen-1.1.3/ViPTreeGen ../GenomeViruses/V-MAGv0.4_fix.fa ../viptree/-MAGv0.4/ --ncpus 64\"
#./qsub_DDBJ.sh medium 8 8 48 bash -c \"source ${HOME}/anaconda3/etc/profile.d/conda.sh\; conda activate py39\; ${HOME}/software/ViPTreeGen-1.1.3/ViPTreeGen ../GenomeViruses/V-MAGv0.4_fix.fa ../viptree/-MAGv0.4/ --ncpus 8\"

#plasmid-MAG------------------------------------------------------------------
#calculate ANI and make distance matrix
mkdir ../fastANI -p
ls ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa > ../fastANI/list_plasmid.tsv
fastANI --ql ../fastANI/list_plasmid.tsv --rl   ../fastANI/list_plasmid.tsv -o ../fastANI/plasmid-MAGv0.4.tsv -t 6


#===================================================================================================================================================================================================
# Analyses of RM system and other phage defence systems  (P-, V-, plasmid-MAG)
#===================================================================================================================================================================================================
#RM genes prediction------------------------------------------------------------------
for f in ../gene/AllGenomev0.6_fix/*.pro.faa; do 
    echo ${f}
    ./fasta_similarity-search.sh -i ${f} -t REBASE -s diamond -T 10
done
mkdir ../blast/P-AllMAGandSAGv0.6_REBASE ../blast/V-MAGv0.4_REBASE ../blast/plasmid-MAGv0.4_REBASE -p
mv ../blast/diamond_P-{MAG,SAG}v0.4_* ../blast/P-AllMAGandSAGv0.6_REBASE/ -f
mv ../blast/diamond_V-MAGv0.4_*       ../blast/V-MAGv0.4_REBASE/ -f
mv ../blast/diamond_plasmid-MAGv0.4_* ../blast/plasmid-MAGv0.4_REBASE/ -f

output=../blast/P-AllMAGandSAGv0.6_REBASE_all.tsv; rm ${output}; for f in ../blast/P-AllMAGandSAGv0.6_REBASE/*; do  
output=../blast/V-MAGv0.4_REBASE_all.tsv;          rm ${output}; for f in ../blast/V-MAGv0.4_REBASE/*;          do          
output=../blast/plasmid-MAGv0.4_REBASE_all.tsv;    rm ${output}; for f in ../blast/plasmid-MAGv0.4_REBASE/*;    do    
    sample=`basename ${f} | rev | cut -f3- -d "." | rev`
    echo ${sample}
    cat ${f} | sed -e "s/^/${sample}\t/g" >> ${output}
done

#RM category summarize for all MAGs
cat   ../blast/P-AllMAGandSAGv0.6_REBASE_all.tsv | cut -f3 | python3 MTase_getinfo_fromEnzymeNameList.py > ../MTase/P-AllMAGandSAGv0.6_REBASE_detail.tsv
cat   ../blast/V-MAGv0.4_REBASE_all.tsv          | cut -f3 | python3 MTase_getinfo_fromEnzymeNameList.py > ../MTase/V-MAGv0.4_REBASE_detail.tsv
cat   ../blast/plasmid-MAGv0.4_REBASE_all.tsv    | cut -f3 | python3 MTase_getinfo_fromEnzymeNameList.py > ../MTase/plasmid-MAGv0.4_REBASE_detail.tsv
paste ../blast/P-AllMAGandSAGv0.6_REBASE_all.tsv ../MTase/P-AllMAGandSAGv0.6_REBASE_detail.tsv           > ../MTase/P-AllMAGandSAGv0.6_REBASE_paste.tsv
paste ../blast/V-MAGv0.4_REBASE_all.tsv          ../MTase/V-MAGv0.4_REBASE_detail.tsv                    > ../MTase/V-MAGv0.4_REBASE_paste.tsv
paste ../blast/plasmid-MAGv0.4_REBASE_all.tsv    ../MTase/plasmid-MAGv0.4_REBASE_detail.tsv              > ../MTase/plasmid-MAGv0.4_REBASE_paste.tsv
cat   ../MTase/{P-AllMAGandSAGv0.6,V-MAGv0.4,plasmid-MAGv0.4}_REBASE_paste.tsv > ../MTase/AllGenomev0.6_REBASE_paste.tsv
cat   ../MTase/AllGenomev0.6_REBASE_paste.tsv | cut -f1,15 | grep -e "\sM" -e "\sRM" | cut -f1 | uniq -c | sed -e "s/diamond_//g" | awk '{print $2"\t"$1}' > ../MTase/AllGenomev0.6_REBASE_countMTase.tsv  
cat   ../MTase/AllGenomev0.6_REBASE_paste.tsv | cut -f1,15 | grep -e "\sR" -e "\sRM" | cut -f1 | uniq -c | sed -e "s/diamond_//g" | awk '{print $2"\t"$1}' > ../MTase/AllGenomev0.6_REBASE_countREase.tsv  

#get MTase sequences
mkdir    ../blast/P-AllMAGandSAGv0.6_REBASE_MTase/
for f in ../blast/P-AllMAGandSAGv0.6_REBASE/*; do
    echo ${f}
    filename=`basename ${f} | sed -e "s/REBASE/REBASE_MTase/g"`
    cat ${f} | grep "\sM\." > ../blast/P-AllMAGandSAGv0.6_REBASE_MTase/${filename}
done

#./getseq_blast_output.sh ../gene/P-MAGsv0.4_all.protein.faa ../blast/P-MAGv0.4_REBASE_MTase.tsv
# diamond_V-MAGv0.4_0th_NSGRY_323.cds_REBASE.tsv  diamond_P-MAGv0.4_2nd_SY_s6_007.protein_REBASE.tsv  diamond_plasmid-MAGv0.4_2nd_NWPT_s67_010.cds_REBASE.tsv
#for f in ../blast/{P-AllMAGandSAGv0.6,V-MAG,plasmid-MAG}v0.4_REBASE/*; do
for f in ../blast/P-AllMAGandSAGv0.6_REBASE_MTase/*; do
    #../blast/V-MAGv0.4_REBASE/diamond_V-MAGv0.4_1st_SY_s6_7_1.cds_REBASEv20240301.tsv
    cds_filename=`basename ${f} | sed -e "s/^diamond_//g" -e "s/_REBASE_MTasev20240301.tsv$/.faa/g" ` ##| sed -e "s/pro/protein/g"
    MAGtype=`     echo ${cds_filename} | cut -f1 -d "_"`  #P-MAGv0.4
    #./getseq_blast_output.sh ../gene/${MAGtype}/${cds_filename} ${f}
     ./getseq_blast_output.sh ../gene/AllGenomev0.6_fix/${cds_filename} ${f}
done

#alphafold2 for some MTases --------------------------------------------------------------------------
mkdir  ../alphafold/query_base -p
for f in ../CandidateSeq_faa/diamond_P-MAGv0.4*.protein.faa; do
    seqkit split -s 1 ${f} -O ../alphafold/query_base
done

#for f in ../alphafold/query_base/*.faa ;  do 
for f in ../alphafold/query_base/diamond_P-MAGv0.4_2nd_NSBL_s6_001*.faa ;  do 
    BASE_FILENAME=` basename ${f} | sed -e "s/.faa//g" `
    if [ -e ../alphafold/work/${BASE_FILENAME}/ranked_0.pdb ]; then
        echo "Structure file already exist. Skip. ${f}"
        continue 
    fi
    #echo 
    ./qsub_ES.sh gpu 1 1 6  ./alphafold2.sh ${f}; 
done

#clean
for f in ../alphafold/work/*; do
    rm -r ${f}/msas
done

#copy
mkdir ../alphafold/summary_pdb/ -p
for f in ../alphafold/work/*; do
    FILENAME=${f##*/}
    echo ${f}
    cp ${f}/ranked_0.pdb  ../alphafold/summary_pdb/${FILENAME}.pdb -n
done

# predic tracts --------------------------------------------------------------------------
conda activate py38
mkdir ../tract/AllGenomev0.6_fix -p
for f in ../GenomeAll/AllGenomev0.6_fix/*.fa; do 
    python3 fasta_ShortSeqeucenRepeat.py ${f}
done
mv  ../tract/tract_* ../tract/AllGenomev0.6_fix
cat ../tract/AllGenomev0.6_fix/tract_* | awk '!a[$0]++'  > ../tract/tract_AllGenomev0.6_fix.tsv

#padloc --------------------------------------------------------------------------
for f in ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/*.fa; do
    filename=` basename ${f} `  #including ".fa"
    if [ -e ../padloc/P-AllMAGandSAGv0.6_fix/${filename}.check ]; then 
        echo "Already exist. Skip: ../padloc/work_DBv2.0.0/${filename}.check"
        continue 
    fi
    echo ${f}
    file_gff=`echo ${f} | sed -e "s/.protein.faa$/.genome.gff/g"`  # using dfast output (NOT WORK. DFAST CHANGED SEQ IDS IN FASTA THAT NOT MATCHED TO GFF)
    #file_gff=`echo ${f} | sed -e "s/.faa$/.gff/g"`  # for virus and plasmid
    
    echo ./qsub_ES.sh cpu 1 1 1 ./padloc.sh genome ${f}  60
    #echo ./qsub_ES.sh cpu 1 1 1 ./padloc.sh gene   ${f} ${file_gff} 60
done
mkdir ../padloc/P-AllMAGandSAGv0.6_fix
mv  ../padloc/P-{MAG,SAG}v0.4_* ../padloc/P-AllMAGandSAGv0.6_fix/
cat ../padloc/P-AllMAGandSAGv0.6_fix/*.csv | grep -v "^system.number" | grep -v "_other" > ../padloc/padloc_P-AllMAGandSAGv0.6_fix_DBv2.0.0.csv


#CRISPR detection --------------------------------------------------------------------------
qlogin -q cpu_I -l elapstim_req=8:00:00 -X

#CRISPRDetect3 
#for f in ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fix/*.fa; do
#for f in ../GenomeViruses/V-MAGv0.4_fix/*.fa; do     ##V-MAGsv0.4_0th_NSBL_14.fa
#for f in ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa; do 
# for f in ../GenomeAll/AllGenomev0.6_fix/*.fa; do 
#     filename=`basename ${f}`
#     nohup     ./CRISPR_CRISPRDetect3.sh  ${f} 10 &
#     sleep 1
# done

#conda activate py38
conda activate py38
parallel -j 8 ./CRISPR_CRISPRDetect3.sh  ${f} 4 ::: ../GenomeAll/AllGenomev0.6_fix/*.fa

mv  ../CRISPR/CRISPRDetect3/ ../CRISPR/CRISPRDetect3_AllGenomev0.6_fix/
cat ../CRISPR/CRISPRDetect3_AllGenomev0.6_fix/*.out | grep "# Summary:" | cut -f4- -d " " | cut -f1-7 -d ";" | sed -e "s/ /\t/g" -e "s/;//g" > ../CRISPR/summary_AllGenomev0.6_info.tsv

# CRISPR_CRISPRCasTyper
#./CRISPR_CRISPRCasTyper.sh ../contig/contig_easyCas10.DSSMv02_Contig_u10k.fasta         meta 30
conda activate py38
      parallel -j 20 ./CRISPR_CRISPRCasTyper.sh  {}  single 4 --load 90% --dry-run ::: ../GenomeAll/AllGenomev0.6_fix/*.fa  &
nohup parallel -j 20 ./CRISPR_CRISPRCasTyper.sh  {}  single 4 --load 90%           ::: ../GenomeAll/AllGenomev0.6_fix/*.fa  &

mkdir ../CRISPR_Tools/CRISPRCasTyper_AllGenomev0.6_fix
mv ../CRISPR_Tools/CRISPRCasTyper_{P-MAGv0.4,P-SAGv0.4,V-MAGv0.4,plasmid-MAGv0.4}* ../CRISPR_Tools/CRISPRCasTyper_AllGenomev0.6_fix

# get figures
mkdir ../CRISPR_Tools/CRISPRCasTyper_AllGenomev0.6_fix_figure
for f in ../CRISPR_Tools/CRISPRCasTyper_AllGenomev0.6_fix/*/plot.png; do
    GenomeID=`echo ${f} | cut -f4 -d "/" | sed -e "s/CRISPRCasTyper_//g"`
    echo ${GenomeID}
    cp ${f} ../CRISPR_Tools/CRISPRCasTyper_AllGenomev0.6_fix_figure/${GenomeID}.png
done


#===================================================================================================================================================================================================
# Host prediction of V-MAG and plasmid-MAG
#===================================================================================================================================================================================================
#taxonomic assignment of Prokaryotic genomes for host prediction
rm ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/*.fai
./qsub_ES.sh gpu 1 1 48 ./gtdbtk.sh de_novo_wf_bacteria ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/ 16   # ~15h
./qsub_ES.sh gpu 1 1 48 ./gtdbtk.sh de_novo_wf_archaea  ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/ 16   # ~1h

#gather
mkdir ../GTDB-Tk/de_novo_wf_Total_P-AllMAGandSAGv0.6_fix_R226
cp ../GTDB-Tk/de_novo_wf_{archaea,bacteria}_P-AllMAGandSAGv0.6_fix_R226/*   ../GTDB-Tk/de_novo_wf_Total_P-AllMAGandSAGv0.6_fix_R226/ -r

#add P-MAGs to the genome database
# may take >30 mins, and will use large RAM size thus reduce thread number
  qlogin -q cpu_I -l elapstim_req=8:00:00 -X
# qlogin -q gpu_S -l elapstim_req=8:00:00 -X
rm ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix_R226/*.fai
conda activate iphop
iphop add_to_db --fna_dir ../GenomeProkaryotes/P-AllMAGandSAGv0.6_fix/ --gtdb_dir ../GTDB-Tk/de_novo_wf_Total_P-AllMAGandSAGv0.6_fix_R226/ --db_dir ${HOME}/database/iPHoP/iPHoP_2023Aug23/Aug_2023_pub_rw --out_dir ${HOME}/database/iPHoP/iPHoP_HotSpringv6_2023Aug23/ --num_threads 60

#iphop 
# This required many RAM space, i.e., >128GB with 60 threads. So, should reduce threads.--------------
# 20 threads under ES GPU node will take ~4h (20240925)
#./qsub_ES.sh   gpu  1 1  24 ./iPHoP.sh ../GenomeViruses/V-MAGv0.4_fix.fa        20 
 ./qsub_ES.sh   gpu  1 1  24 ./iPHoP.sh ../GenomeViruses/V-MAGv0.4_fix.fa        20 ${HOME}/database/iPHoP/iPHoP_HotSpringv6_2023Aug23
#./qsub_ES.sh   cpu  3 1  24 ./iPHoP.sh ../GenomeViruses/V-MAGv0.4_fix.fa        60 ${HOME}/database/iPHoP/iPHoP_HotSpringv6_2023Aug23
#./qsub_ES.sh   gpu  1 1  24 ./iPHoP.sh ../GenomePlasmids/plasmid-MAGv0.4_fix.fa 20 
 ./qsub_ES.sh   gpu  1 1  24 ./iPHoP.sh ../GenomePlasmids/plasmid-MAGv0.4_fix.fa 20 ${HOME}/database/iPHoP/iPHoP_HotSpringv6_2023Aug23
rm ../HostPrediction/iPHoP_*/Wdir -r

#blast alignment for P-MAGs vs plasmid-MAGs ----------------
makeblastdb ../GenomeProkaryotes/P-MAGv0.4_fix.fa
blastn --db ../GenomeProkaryotes/P-MAGv0.4_fix.blast --query ../GenomePlasmids/plasmid-MAGv0.4_fix.fa   --outfmt 6  --threads 10 --out ../blast/P-MAGv0.4_plasmid-MAGv0.4.tsv


#===================================================================================================================================================================================================
# Modification detection and motif prediction (naive mapping strategy)
#===================================================================================================================================================================================================
# # BAM file downsampling using samtools--------------------------------------------------
# for f in ../mapping/*.mask/pbmm2HiFi_*.hifi_kinetics_sorted.bam; do
#     filename=`basename ${f} | rev | cut -f2- -d "." | rev`
#     coverage=`cat ../SummaryInformation/MAG_coverage.txt | grep ${filename} | cut -f7 -d " "`
#     echo "    ${filename}: ${coverage}"
#     frac=` echo "scale=3;  200 / ${coverage}  " | bc`
#     echo "    coverage: ${coverage} , frac: ${frac}"
#     if [ `echo "$frac > 0.9" | bc` == 1 ]; then
#         continue
#     fi
#
#     new_filename=`echo ${f} | sed -e "s/sorted.bam/sorted.downsamplingCov200.bam/g"`
#     echo "Down scale: ${f} -> ${new_filename}"  
#     echo 
#     #./qsub_ES.sh cpu 1 1 2 
#     samtools view -b -s ${frac} ${f} -o ${new_filename} -@ 30
# done
#
# #make index
# for f in ../checkm_Genome/CheckM2_P-MAGv0.4_masked_chromosome/*.fa; do
#     if [ ! -e ${f}.fai ]; then
#         echo "make fai file: ${f}"
#         samtools faidx ${f}
#     fi
# done
#
# #check
# for f in *.check; do
#     basemodgff=`echo ${f} | sed -e "s/.check/.gff/g"`
#     if [ ! -e ${basemodgff} ]; then
#         echo "No File: :"${basemodgff}
#     fi
# done

# # BAM file downsampling using bamtools--------------------------------------------------
# #make BAM index
# qlogin -q cpu_I -l elapstim_req=8:00:00 -X  
# mkdir ../coverage/alignedReadsToEachMAG_elprep2bamtools_downsampling1/
# mkdir ../coverage/alignedReadsToEachMAG_elprep2bamtools_downsampling2/
# for f in ../GenomeProkaryotes/P-MAGv0.4_fixmask/*.fa;  do     # P-MAGv0.4_1st_NSBL_Circle.part_005.mask.fa
# #for f in ../GenomeViruses/V-MAGv0.4_fixmask/*.fa; do     # V-MAGsv0.4_0th_NSBL_14.mask.fa
# #for f in ../GenomePlasmids/plasmid-MAGv0.4_fixmask/*.fa; do   
#     echo ${f}
#     MAGname=`       basename ${f} | sed -e "s/.fa$//g" -e "s/.mask//g"` 
#     sample=`        basename ${f} | cut -f3 -d "_"` 
#     samtools index -@ 60 ../coverage/alignedReadsToEachMAG_elprep2bamtools/alignedfixmask_${MAGname}_${sample}.bam
# done

# #downsampling
# conda activate java   # allow to use java at qsub nodes in ES system
# for f in ../GenomeProkaryotes/P-MAGv0.4_fixmask/*.fa;  do     # P-MAGv0.4_1st_NSBL_Circle.part_005.mask.fa
#     echo ${f}
#     MAGname=`       basename ${f} | sed -e "s/.fa$//g" -e "s/.mask//g"`  
#     sample=`        basename ${f} | cut -f3 -d "_"` 

#     #check read count
#     #bamtools count -in ../coverage/alignedReadsToEachMAG_elprep2bamtools/alignedfixmask_${MAGname}_${sample}.bam 
#     # if [ -e ../coverage/alignedReadsToEachMAG_elprep2bamtools_downsampling1/alignedfixmaskdownsampling_${MAGname}_${sample}.bam ]; then 
#     #     continue 
#     # fi
    
#     #downsampling
#     #./qsub_ES.sh cpu 1 1 6  \
#     #    bash -c \"${HOME}//local/bin/bamtools random -in                ../coverage/alignedReadsToEachMAG_elprep2bamtools/alignedfixmask_${MAGname}_${sample}.bam \
#     #                    -out ../coverage/alignedReadsToEachMAG_elprep2bamtools_downsampling1/alignedfixmaskdownsampling_${MAGname}_${sample}.bam -n 20000\"   #VERY SLOW
    
#     #setting1: 20000 reads/P-MAGs
#     # nohup ${HOME}/software/bbmap/reformat.sh samplereadstarget=20000 \
#     #     in=../coverage/alignedReadsToEachMAG_elprep2bamtools/alignedfixmask_${MAGname}_${sample}.bam \
#     #     out=../coverage/alignedReadsToEachMAG_elprep2bamtools_downsampling1/alignedfixmaskdownsampling_${MAGname}_${sample}.bam   &   #

#     #setting2: 200 Mb/P-MAGs
#     nohup ${HOME}/software/bbmap/reformat.sh samplebasestarget=200000000 \
#         in=../coverage/alignedReadsToEachMAG_elprep2bamtools/alignedfixmask_${MAGname}_${sample}.bam \
#         out=../coverage/alignedReadsToEachMAG_elprep2bamtools_downsampling2/alignedfixmaskdownsampling2_${MAGname}_${sample}.bam   &   #20000 reads/P-MAGs

#     sleep 1
# done

# for f in ../{GenomeViruses,GenomePlasmids}/{V,plasmid}-MAGv0.4_fixmask/*.fa; do     # V-MAGsv0.4_0th_NSBL_14.mask.fa
#     echo ${f}
#     MAGname=`       basename ${f} | sed -e "s/.fa$//g" -e "s/.mask//g"`  
#     sample=`        basename ${f} | cut -f3 -d "_"` 
#     if [ -e ../coverage/alignedReadsToEachMAG_elprep2bamtools_downsampling1/alignedfixmaskdownsampling_${MAGname}_${sample}.bam ]; then 
#         continue 
#     fi
#     nohup ${HOME}/software/bbmap/reformat.sh samplereadstarget=1000 \
#         in=../coverage/alignedReadsToEachMAG_elprep2bamtools/alignedfixmask_${MAGname}_${sample}.bam \
#         out=../coverage/alignedReadsToEachMAG_elprep2bamtools_downsampling1/alignedfixmaskdownsampling_${MAGname}_${sample}.bam   &   #1000 reads/V,plasmid-MAGs

#     sleep 1
# done

#modification detection (Naive)--------------------------------------------------
rm tmp_modification_2ndRun.sh tmp_modification_2ndRun_*.sh
rm tmp_modification_3rdRun.sh

#for f in ../GenomeProkaryotes/P-MAGv0.4_fix/*.fa;    do
#for f in ../GenomeProkaryotes/P-MAGv0.4_fixmask/*.fa;    do
for f in ../singlecell/P-SAGv0.4_fixmask/*.fa;   do
#for f in ../GenomeViruses/V-MAGv0.4_fix/*.fa;   do 
#for f in ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa; do 
#for f in ../GenomeProkaryotes/P-AllMAGandSAGv0.4/*.fa;   do
    echo ${f}
    GenomeID=`    basename ${f} | sed -e "s/.fa//g" -e "s/.mask//g" ` # P-MAGv0.4_1st_NSGRN_Circle_021
    GenomeIDmask=`basename ${f} | sed -e "s/.fa//g" `                 # P-MAGv0.4_1st_NSGRN_Circle_021.mask
    sample=`echo ${GenomeID} | cut -f3 -d "_" `                       # NSGRN

    #--------------------------------------------------------
    # direct mapping
    File_map="         ../mapping/${GenomeIDmask}/pbmm2HiFi_${GenomeIDmask}_${sample}.hifi_kinetics_sorted.bam"            #../mapping/1st_DW.part_002/pbmm2HiFi_1st_DW.part_002_DW.hifi_kinetics_sorted.bam
    File_csv="                 ../modification/v25pbmm2HiFi_${GenomeIDmask}_${sample}.hifi_kinetics_sorted.motifs.csv"     #pbmm2HiFi_1st_DW_Circle.part_014_DW.hifi_kinetics_sorted.basemods.csv
    #File_mappingDOWN="../mapping/${GenomeIDmask}/pbmm2HiFi_${GenomeIDmask}_${sample}.hifi_kinetics_sorted.downsamplingCov200.bam"
    #File_csvDOWN="               ../modification/pbmm2HiFi_${GenomeIDmask}_${sample}.hifi_kinetics_sorted.downsamplingCov200.motifs.csv"  
    #--------------------------------------------------------
    if [ -e ${File_csv} ]  ; then  #|| [ -e ${File_csv2}  ] 
        echo "Already finished. Skip: ${File_csv}"
        continue
    fi

    if [ ! -e ${File_map} ]; then 
        echo "No mapping file exist: ${File_map}"; 
        continue; 
    fi

    #REQUIRE SOME AMOUNT OF ROM SPACE (>120GB in some case with 120 threads)     
    echo ./qsub_ES.sh cpu 1 1 6 ./modification_pacbio.sh ${f} ${File_map} 60  v25  #m6A, m5C, and m4C  #1st run

    #2nd run
    #echo "nohup  ./modification_pacbio.sh ${f} ${File_map} 60  v25  &"  >>  tmp_modification_2ndRun.sh #m6A, m5C, and m4C  

    #3rd run
    #echo     
    #./qsub_ES.sh cpu 1 1 48   ./modification_pacbio.sh ${f} ${File_map} 60     v25  #343/373 files were completed even run this command
done

# #2nd run
# split -l 8 --additional-suffix=.sh tmp_modification_2ndRun.sh   tmp_modification_2ndRun_
# chmod +x tmp_modification_2ndRun_*
# for f in tmp_modification_2ndRun_*; do
#     ./qsub_ES.sh cpu 3 1  48 ./${f}
# done

# #3rd run
# split -l 4 --additional-suffix=.sh tmp_modification_3rdRun.sh   tmp_modification_3rdRun_
# chmod +x tmp_modification_3rdRun_*
# for f in tmp_modification_3rdRun_*; do
#     ./qsub_ES.sh cpu 3 1  48 ./${f}
# done

# find ../modification/*.gff  -empty -delete
# find ../modification/*.csv  -empty -delete

# # rescure #1
# ./qsub_ES.sh cpu 3 1 48  ./modification_wrapper1.sh
# ./qsub_ES.sh gpu 1 4 48  ./modification_wrapper1.sh
# ./qsub_ES.sh gpu 1 4 48  ./modification_wrapper2.sh

# # rescure #2
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 1st_DW
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 1st_NSBL
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 1st_NSBR
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 1st_NSGF
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 1st_NSGRN
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 1st_NSGRY
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 1st_NSOL
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 1st_NWPT
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 1st_NWRS
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 1st_SY

# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 2nd_DW
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 2nd_NSBL
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 2nd_NSBR
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 2nd_NSGF
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 2nd_NSGRN
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 2nd_NSGRY
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 2nd_NSOL
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 2nd_NWPT
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 2nd_NWRS
# ./qsub_ES.sh gpu 1 1 48  ./modification_wrapper.sh 2nd_SY


#===================================================================================================================================================================================================
# Modification detection and motif prediction (map-sepalate-remap strategy)
#===================================================================================================================================================================================================
# # mapping the HiFi read to the same-sourced MAGs again--------------------------------------------------
# #  (reads that from the same sample to ones which used for the assembling and MAG definition)
# #for f in ../GenomeProkaryotes/P-MAGv0.4_fix/*.fa;     do    # P-MAGv0.4_1st_NSBL_Circle.part_005.fa
# for f in ../GenomeProkaryotes/P-MAGv0.4_fixmask/*.fa;  do    # P-MAGv0.4_1st_NSBL_Circle.part_005.mask.fa
# #for f in ../singlecell/P-SAGv0.4_fixmask/*.fa;        do
# #for f in ../GenomeViruses/V-MAGv0.4_fix/*.fa;         do    # V-MAGsv0.4_0th_NSBL_14.fa
# #for f in ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa;  do    
# #for f in ../GenomeProkaryotes/P-AllMAGandSAGv0.4/*.fa; do
#     GenomeID=`    basename ${f} | sed -e "s/.fa$//g" -e "s/.mask//g"` 
#     GenomeIDmask=`basename ${f} | sed -e "s/.fa$//g"` 
#     sample=`      basename ${f} | cut -f3 -d "_"` 
#     if [ -f ../mapping/${GenomeIDmask}/pbmm2HiFi_${GenomeIDmask}_alignedfixmask_${GenomeID}_${sample}_sorted.bam.check ]; then  
#         echo "Already exist. Skip"
#         continue
#     fi
#     echo ./qsub_ES.sh cpu 1 1 6  ./mapping.sh -t pbmm2HiFi -b ${f} -i ../coverage/alignedReadsToEachMAG_elprep2bamtools/alignedfixmask_${GenomeID}_${sample}.bam -T 64
# done

# #modification detection--------------------------------------------------
# rm tmp_modification_2ndRun.sh tmp_modification_2ndRun_*.sh
# rm tmp_modification_3rdRun.sh

# #for f in ../GenomeProkaryotes/P-MAGv0.4_fix/*.fa;    do
# for f in ../singlecell/P-SAGv0.4_fixmask/*.fa;   do
# #for f in ../GenomeViruses/V-MAGv0.4_fix/*.fa;   do 
# #for f in ../GenomePlasmids/plasmid-MAGv0.4_fix/*.fa; do 
# #for f in ../GenomeProkaryotes/P-MAGv0.4_fixmask/*.fa;    do
# #for f in ../GenomeProkaryotes/P-AllMAGandSAGv0.4/*.fa;   do
#     GenomeID=`    basename ${f} | sed -e "s/.fa//g" -e "s/.mask//g" `    #P-MAGv0.4_1st_NSGRN_Circle_021
#     GenomeIDmask=`basename ${f} | sed -e "s/.fa//g" `                    #P-MAGv0.4_1st_NSGRN_Circle_021.mask
#     sample=`   echo ${GenomeID} | cut -f3 -d "_" ` 

#     #--------------------------------------------------------
#     #re-mapped reads
#     #P-MAGv0.4_1st_NSOL_Circle_002/pbmm2HiFi_P-MAGv0.4_1st_NSOL_Circle_002_aligned_P-MAGv0.4_1st_NSOL_Circle_002_NSOL_sorted.bam
#     #v12defaultpbmm2HiFi_P-MAGv0.4_2nd_SY_s65_001_aligned_P-MAGv0.4_2nd_SY_s65_001_SY_sorted.motifs.csv
#     File_map="           ../mapping/${GenomeIDmask}/pbmm2HiFi_${GenomeIDmask}_alignedfixmask_${GenomeID}_${sample}_sorted.bam"  
#     File_csv="                   ../modification/v12pbmm2HiFi_${GenomeIDmask}_alignedfixmask_${GenomeID}_${sample}_sorted.motifs.csv"  
#     File_csv2="../modification/P-MAGv0.4_v12mask/v12pbmm2HiFi_${GenomeIDmask}_alignedfixmask_${GenomeID}_${sample}_sorted.motifs.csv"

#     #--------------------------------------------------------
#     if [ -e ${File_csv} ]  ; then  #|| [ -e ${File_csv2} ] 
#         echo "Already finished. Skip: ${File_csv}"
#         continue
#     fi

#     if [ ! -e ${File_map} ]; then 
#         echo "No mapping file exist: ${File_map}"; 
#         continue; 
#     fi

#     #REQUIRE SOME AMOUNT OF ROM SPACE (>120GB in some case with 120 threads)     
#     ./qsub_ES.sh cpu 1 1  6   ./modification_pacbio.sh ${f} ${File_map} 60  v12  #m6A, m5C, and m4C  #1st run

#     #2nd run
#     #echo ${f}
#     #echo "nohup  ./modification_pacbio.sh ${f} ${File_map} 60  v12  &"  >>  tmp_modification_2ndRun.sh #m6A, m5C, and m4C  

#     #3rd run
#     echo ${f}
#     #echo "nohup  ./modification_pacbio.sh ${f} ${File_map} 300  v12  &"  >>  tmp_modification_3rdRun.sh #m6A, m5C, and m4C 
#     #echo     ./qsub_ES.sh cpu 1 1  48   ./modification_pacbio.sh ${f} ${File_map} 60  v12  #343/373 files were completed even run this command
# done

# #2nd run
# split -l 8 --additional-suffix=.sh tmp_modification_2ndRun.sh   tmp_modification_2ndRun_
# chmod +x tmp_modification_2ndRun_*
# for f in tmp_modification_2ndRun_*; do
#     ./qsub_ES.sh cpu 3 1  48 ./${f}
# done

# #3rd run
# split -l 4 --additional-suffix=.sh tmp_modification_3rdRun.sh   tmp_modification_3rdRun_
# chmod +x tmp_modification_3rdRun_*
# for f in tmp_modification_3rdRun_*; do
#     ./qsub_ES.sh cpu 3 1  48 ./${f}
# done

# find ../modification/*.gff  -empty -delete
# find ../modification/*.csv  -empty -delete


#===================================================================================================================================================================================================
# Modification detection and motif prediction (map-separate strategy)
# motif prediction is completed if the last line in log file is like "[INFO] 2025-08-19 00:40:01,071Z [kineticsTools.ipdSummary _pacbio_main_runner 207] exiting with return code 0 in 129332.07 sec."
#===================================================================================================================================================================================================
# modification detectin using genome-separated BAM files 
for f in ../singlecell/P-SAGv0.4_fixmask/*.fa;   do  #***
#for f in ../GenomeAll/AllGenomev0.6_fixmask/P-{M,S}AGv0.4_*.mask.fa;    do  #***
#for f in ../GenomeAll/AllGenomev0.6_fixmask/V-MAGv0.4_*.mask.fa;       do  #*** 
#for f in ../GenomeAll/AllGenomev0.6_fixmask/plasmid-MAGv0.4_*.mask.fa; do  #*** 
#for f in ../GenomeAll/AllGenomev0.6_fixmask/P-MAGv0.4_2nd_NSBR_*.mask.fa; do  # ReRun: 20250618
    MAGname=`       basename ${f}   | sed -e "s/.fa$//g" -e "s/.mask//g"` #P-MAGv0.4_1st_NSGRN_Circle_021
    MAGnameMapping=`basename ${f}   | sed -e "s/.fa$//g" `                #P-MAGv0.4_1st_NSGRN_Circle_021.mask
    dataset=`       dirname  ${f}   | rev | cut -f1 -d "/" | rev`         #AllGenomev0.5_fixmask
    datasetPrefix=` echo ${dataset} | cut -f1 -d "_"`                     #AllGenomev0.5
    sample=`        echo ${MAGname} | cut -f3 -d "_" `                    #NSGRN
    mode=v25
    #mappingDir=pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted  #P-SAG
    mappingDir=pbmm2HiFi_${dataset}_TOTAL  #AllGenome
    
    # in case for AllGenome
    #File_map="../ReadSeparation/${mappingDir}/pbmm2HiFi_${dataset}_${sample}.hifi_kinetics_sorted_${MAGname}.bam"  
    File_map="../ReadSeparation/${mappingDir}/MapSep-pbmm2HiFi_${dataset}_${sample}.hifi_kinetics_sorted_${MAGname}.bam"  

    # in case for P-SAG
    #File_map="../ReadSeparation/pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted/MapSep-pbmm2HiFi_${dataset}_${sample}.hifi_kinetics_sorted_${MAGname}.bam" 

    #File_check="../modification/v25MapSep-pbmm2HiFi_${dataset}_${sample}.hifi_kinetics_sorted_${MAGname}.basemods.check"
    #v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_221.motifs.csv
    File_check1_1="../modification/${mode}MapSep-pbmm2HiFi_${dataset}_${sample}.hifi_kinetics_sorted_${MAGname}.motifs.csv"
    #File_check1_2="../modification/${datasetPrefix}_${mode}maskMapSep/original/v25MapSep-pbmm2HiFi_${dataset}_${sample}.hifi_kinetics_sorted_${MAGname}.motifs.csv"
    File_check1_2="../modification/${datasetPrefix}_${mode}maskMapSep/original/v25MapSep-pbmm2HiFi_${dataset}_${sample}.hifi_kinetics_sorted_${MAGname}.motifs.csv"
    File_check1_3="../modification/${datasetPrefix}_${mode}maskMapSep/original_1stES/v25MapSep-pbmm2HiFi_${dataset}_${sample}.hifi_kinetics_sorted_${MAGname}.motifs.csv"
    File_check1_4="../modification/${datasetPrefix}_${mode}maskMapSep/original_1stES_motifs/v25MapSep-pbmm2HiFi_${dataset}_${sample}.hifi_kinetics_sorted_${MAGname}.motifs.csv"
    File_check1_5="../modification/${datasetPrefix}_${mode}maskMapSep/original_2ndDDBJ/v25MapSep-pbmm2HiFi_${dataset}_${sample}.hifi_kinetics_sorted_${MAGname}.motifs.csv"
    File_check2_1="../modification/${datasetPrefix}_${mode}maskMapSep/rename/${MAGname}.basemods.gff"
    if [ -e ${File_check1_1} ] ; then echo "    Already completed: File_check1_1. Skip: ${f}"; continue; fi  # check
    if [ -e ${File_check1_2} ] ; then echo "    Already completed: File_check1_2. Skip: ${f}"; continue; fi  # check
    if [ -e ${File_check1_3} ] ; then echo "    Already completed: File_check1_3. Skip: ${f}"; continue; fi  # check
    if [ -e ${File_check1_4} ] ; then echo "    Already completed: File_check1_4. Skip: ${f}"; continue; fi  # check
    if [ -e ${File_check1_5} ] ; then echo "    Already completed: File_check1_5. Skip: ${f}"; continue; fi  # check
    #if [ -e ${File_check2}   ] ; then echo "    Output gff exist.  Skip: ${f}"; continue; fi  # gff
    #v25MapSep-pbmm2HiFi_AllGenomev0.4_fixmask_NSBL.hifi_kinetics_sorted_P-MAGv0.4_1st_NSBL_009.basemods.check
    #echo ${File_check1_2}
    
    # 1th ES run round 1
    # 297/373 (P-MAG in AllGenome) and 194/224 (P-SAG in AllGenome) files were completed when run this command (v0.4)
    #./qsub_ES.sh     cpu  1 1   6 ./modification_pacbio.sh ${f} ${File_map} 64 v25  #***
    
    # 1th ES run round 2
    #./qsub_ES.sh     cpu  3 1 24 ./modification_pacbio.sh ${f} ${File_map} 128 v25  #***
    #echo "./modification_pacbio.sh ${f} ${File_map} 64 v25 &" >> tmp_modification_${datasetPrefix}_${mode}maskMapSep.tsv

    # 2nd DDBJ run
    echo ./sbatch_DDBJ.sh epyc 6 12 360 ./modification_pacbio.sh ${f} ${File_map} 6 v25

    # m4C and m6A only (not use)
    #echo ./qsub_ES.sh     cpu  1 1  4 ./modification_pacbio.sh ${f} ${File_map} 60 defv25  # XXX/XXX (P-MAG) files were completed when run this command
    #echo ./sbatch_DDBJ.sh epyc 4 8 96 ./modification_pacbio.sh ${f} ${File_map}  4 defv25  # XXX/XXX (P-MAG) files were completed when run this command

    #break
done
#./modification_pacbio.sh ../GenomeProkaryotes/P-AllMAGandSAGv0.4_fixmask/P-SAGv0.4_xxx_NSOL_220.mask.fa ../ReadSeparation/pbmm2HiFi_P-SAGv0.4_fixmask_NSGF.hifi_kinetics_sorted_random0.01/P-SAGv0.4_xxx_NSOL_220.bam 6  v12 

#split list for 2nd run
dataset=P-SAGv0.4_v25maskMapSep

rm       tmp_modification_${dataset}_partition_*
split    tmp_modification_${dataset}.tsv -l 3  tmp_modification_${dataset}_partition_ --additional-suffix=.sh
rm       tmp_modification_${dataset}.tsv
chmod +x tmp_modification_${dataset}_partition_*.sh
for f in tmp_modification_${dataset}_partition_*.sh; do
    ./qsub_ES.sh cpu  3 1 24 ./${f}
done
# ./qsub_ES.sh cpu  3 1 24 parallel -j 10 ./virsorter2.sh {} 60 ::: ../assembly/1st_split/*split/hifiasmMeta_*.fa

# easy view
for f in  ../modification/v25MapSep-*_P-SAGv0.4_*motifs.csv; do echo ""; basename ${f}; cat ${f} | grep -v "motifString"; done | less
for f in  ../modification/*V-MAG*motifs.csv;                 do echo ""; basename ${f}; cat ${f} | grep -v "motifString"; done | less
for f in  ../modification/*motifs.csv;                       do echo ""; basename ${f}; cat ${f} | grep -v "motifString"; done | less

# easy view of calculation time
for f in *.log; do
    filename=`basename ${f}`
    time=`cat ${f} | grep "exiting with return code 0" | rev | cut -f2 -d " " | rev`  #[INFO] 2025-08-19 00:40:01,071Z [kineticsTools.ipdSummary _pacbio_main_runner 207] exiting with return code 0 in 129332.07 sec.
    echo -e "${f}\t${time}\ts"
done

# list up completed file
for f in ../modification/AllGenomev0.6_v25maskMapSep/original_*motifs/*P-{SAG,MAG}*motifs.csv; do filename=`basename ${f}`; echo -e "${filename}"; done | sort

# file move (1th and 2nd runs)----------------------------------------------
#GenomeAll ES
mkdir ../modification/AllGenomev0.6_v25maskMapSep/original_1stES/        -p
mkdir ../modification/AllGenomev0.6_v25maskMapSep/original_1stES_motifs/ -p
mv    ../modification/v25MapSep-pbmm2HiFi_AllGenomev0.6_fixmask_*            ../modification/AllGenomev0.6_v25maskMapSep/original_1stES/
cp    ../modification/AllGenomev0.6_v25maskMapSep/original_1stES/*motifs.csv ../modification/AllGenomev0.6_v25maskMapSep/original_1stES_motifs/

#GenomeAll DDBJ
mkdir ../modification/AllGenomev0.6_v25maskMapSep/original_2ndDDBJ/        -p
mkdir ../modification/AllGenomev0.6_v25maskMapSep/original_2ndDDBJ_motifs/ -p
mv    ../modification/v25MapSep-pbmm2HiFi_AllGenomev0.6_fixmask_*              ../modification/AllGenomev0.6_v25maskMapSep/original_2ndDDBJ/
cp    ../modification/AllGenomev0.6_v25maskMapSep/original_2ndDDBJ/*motifs.csv ../modification/AllGenomev0.6_v25maskMapSep/original_2ndDDBJ_motifs/

#P-SAG ES
mkdir ../modification/P-SAGv0.4_v25maskMapSep/original_1stES/         -p
mkdir ../modification/P-SAGv0.4_v25maskMapSep/original_1stES_motifs/  -p
mv    ../modification/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_*            ../modification/P-SAGv0.4_v25maskMapSep/original_1stES/
cp    ../modification/P-SAGv0.4_v25maskMapSep/original_1stES/*motifs.csv ../modification/P-SAGv0.4_v25maskMapSep/original_1stES_motifs/

#P-SAG DDBJ
mkdir ../modification/P-SAGv0.4_v25maskMapSep/original_2ndDDBJ/        -p
mkdir ../modification/P-SAGv0.4_v25maskMapSep/original_2ndDDBJ_motifs/ -p
mv    ../modification/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_* ../modification/P-SAGv0.4_v25maskMapSep/original_2ndDDBJ/
cp    ../modification/P-SAGv0.4_v25maskMapSep/original_2ndDDBJ/*motifs.csv                                 ../modification/P-SAGv0.4_v25maskMapSep/original_2ndDDBJ_motifs/

# merge (do not change order) 
mkdir ../modification/AllGenomev0.6_v25maskMapSep/original/ 
cp    ../modification/AllGenomev0.6_v25maskMapSep/original_1stES/*   ../modification/AllGenomev0.6_v25maskMapSep/original/
cp    ../modification/AllGenomev0.6_v25maskMapSep/original_2ndDDBJ/* ../modification/AllGenomev0.6_v25maskMapSep/original/

mkdir ../modification/P-SAGv0.4_v25maskMapSep/original/      
cp    ../modification/P-SAGv0.4_v25maskMapSep/original_1stES/*       ../modification/P-SAGv0.4_v25maskMapSep/original/
cp    ../modification/P-SAGv0.4_v25maskMapSep/original_2ndDDBJ/*     ../modification/P-SAGv0.4_v25maskMapSep/original/

#file move (each pattern) ----------------------------------------------
#  mkdir -p ../modification/P-MAGv0.4_v12mask{Naive,MapSep}/{original,rename}
#  mkdir -p ../modification/P-SAGv0.4_v12mask{Naive,MapSep}/{original,rename}
# #mkdir -p ../modification/V-MAGv0.4_v12mask{Naive,MapSep}/{original,rename}
# #mkdir -p ../modification/plasmid-MAGv0.4_v12mask{Naive,MapSep}/{original,rename}
#  mkdir -p ../modification/AllGenomev0.4_v12mask{Naive,MapSep}/{original,rename}
#  mkdir -p ../modification/AllGenomev0.5_v12mask{Naive,MapSep}/{original,rename}

#  mkdir -p ../modification/P-MAGv0.4_v25mask{Naive,MapSep}/{original,rename}
#  mkdir -p ../modification/P-SAGv0.4_v25mask{Naive,MapSep}/{original,rename}
# #mkdir -p ../modification/V-MAGv0.4_v25mask{Naive,MapSep}/{original,rename}
# #mkdir -p ../modification/plasmid-MAGv0.4_v25mask{Naive,MapSep}/{original,rename}
#  mkdir -p ../modification/AllGenomev0.4_v25mask{Naive,MapSep}/{original,rename}
#  mkdir -p ../modification/AllGenomev0.5_v25mask{Naive,MapSep}/{original,rename}

# #mkdir -p ../modification/AllGenomev0.4_defv25mask{Naive,MapSep}/{original,rename}

#  #Naive
#  mv ../modification/v12pbmm2HiFi_P-MAGv0.4_fixmask_*               ../modification/P-MAGv0.4_v12maskNaive/original
#  mv ../modification/v12pbmm2HiFi_P-SAGv0.4_fixmask_*               ../modification/P-SAGv0.4_v12maskNaive/original
# # mv ../modification/v12pbmm2HiFi_V-MAGv0.4_fixmask_*               ../modification/V-MAGv0.4_v12maskNaive/original
# # mv ../modification/v12pbmm2HiFi_plasmid-MAGv0.4_fixmask_*         ../modification/plasmid-MAGv0.4_v12maskNaive/original
#  mv ../modification/v12pbmm2HiFi_AllGenomev0.4_fixmask_*           ../modification/AllGenomev0.4_v12maskNaive/original
#  mv ../modification/v12pbmm2HiFi_AllGenomev0.5_fixmask_*           ../modification/AllGenomev0.5_v12maskNaive/original

#  mv ../modification/v25pbmm2HiFi_P-MAGv0.4_fixmask_*               ../modification/P-MAGv0.4_v25maskNaive/original
#  mv ../modification/v25pbmm2HiFi_P-SAGv0.4_fixmask_*               ../modification/P-SAGv0.4_v25maskNaive/original
# #mv ../modification/v25pbmm2HiFi_V-MAGv0.4_fixmask_*               ../modification/V-MAGv0.4_v25maskNaive/original
# #mv ../modification/v25pbmm2HiFi_plasmid-MAGv0.4_fixmask_*         ../modification/plasmid-MAGv0.4_v25maskNaive/original
#  mv ../modification/v25pbmm2HiFi_AllGenomev0.4_fixmask_*           ../modification/AllGenomev0.4_v25maskNaive/original
#  mv ../modification/v25pbmm2HiFi_AllGenomev0.5_fixmask_*           ../modification/AllGenomev0.5_v25maskNaive/original

# #Mapped-read separation
#  mv ../modification/v12MapSep-pbmm2HiFi_P-MAGv0.4_fixmask_*        ../modification/P-MAGv0.4_v12maskMapSep/original
#  mv ../modification/v12MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_*        ../modification/P-SAGv0.4_v12maskMapSep/original
# #mv ../modification/v12MapSep-pbmm2HiFi_V-MAGv0.4_fixmask_*        ../modification/V-MAGv0.4_v12maskMapSep/original
# #mv ../modification/v12MapSep-pbmm2HiFi_plasmid-MAGv0.4_fixmask_*  ../modification/plasmid-MAGv0.4_v12maskMapSep/original
#  mv ../modification/v12MapSep-pbmm2HiFi_AllGenomev0.4_fixmask_*    ../modification/AllGenomev0.4_v12maskMapSep/original
#  mv ../modification/v12MapSep-pbmm2HiFi_AllGenomev0.5_fixmask_*    ../modification/AllGenomev0.5_v12maskMapSep/original

#  mv ../modification/v25MapSep-pbmm2HiFi_P-MAGv0.4_fixmask_*        ../modification/P-MAGv0.4_v25maskMapSep/original
#  mv ../modification/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_*        ../modification/P-SAGv0.4_v25maskMapSep/original
# #mv ../modification/v25MapSep-pbmm2HiFi_V-MAGv0.4_fixmask_*        ../modification/V-MAGv0.4_v25maskMapSep/original
# #mv ../modification/v25MapSep-pbmm2HiFi_plasmid-MAGv0.4_fixmask_*  ../modification/plasmid-MAGv0.4_v25maskMapSep/original
#  mv ../modification/v25MapSep-pbmm2HiFi_AllGenomev0.4_fixmask_*    ../modification/AllGenomev0.4_v25maskMapSep/original
#  mv ../modification/v25MapSep-pbmm2HiFi_AllGenomev0.5_fixmask_*    ../modification/AllGenomev0.5_v25maskMapSep/original

# #re-prediction of motif with manual settings --------------------------------------------------
# MotifMaker=" ${HOME}/software/smrtlink_12.0.0.177059/smrtcmds/bin/motifMaker"
# for f in ../GenomeProkaryotes/P-MAGv0.4_fix/*.fa; do 
#     #1st_NSBL_Circle.part_005.mask.fa
#     filename=`basename ${f} | sed -e "s/.fa//g" -e "s/.mask//g" `         
#     number=`  echo ${filename} | cut -f2 -d "." | cut -f2 -d "_"  `  #001
#     sample=`  echo ${filename} | cut -f1 -d "." | sed -e "s/_Circle//g" | rev | cut -f1 -d "_" | rev `  #NSBL

#     echo ./qsub_ES.sh cpu 1 1 3  \" ${MotifMaker} find -f ${f} -g ../modification/P-MAGv0.3/pbmm2HiFi_${filename}_${sample}.hifi_kinetics_sorted.basemods.gff -o ../modification/P-MAGv0.3/pbmm2HiFi_${filename}_${sample}.hifi_kinetics_sorted.motifs3.csv --minScore 30  -j 60 \" 
# done

#file rename and gather --------------------------------------------------
Dataset=AllGenomev0.6  #*
Dataset=P-SAGv0.4      #*

strategy=v12maskNaive
strategy=v12maskMapSep 
strategy=v25maskNaive
strategy=v25maskMapSep  #*

mkdir ../modification/${Dataset}_${strategy}/rename/
for f in ../modification/${Dataset}_${strategy}/original/*basemods.gff; do      #v12defaultpbmm2HiFi_V-MAGv0.4_1st_SY_s6_12_1_alignedfixmask_V-MAGv0.4_1st_SY_s6_12_1_SY_sorted.basemods.csv
    #new_filename=`basename ${f} | sed -e "s/.hifi_kinetics_sorted.basemods.gff//g" -e "s/pbmm2HiFi_//g" -e "s/v12default//g" | rev | cut -f2- -d "_" | rev`;  #2nd_NSGRY_s226_NSGRY.part_001_NSGRY 
    #new_filename=`basename ${f} | sed -e "s/_sorted.basemods.gff//g" -e "s/pbmm2HiFi_//g" -e "s/${strategy}//g" | cut -f 1-2 -d "." | rev | cut -f3- -d "_" | rev`;  #V-MAGv0.4_1st_SY_s6_12_1 

    # v12pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_224.basemods.gff ->  P-SAGv0.4_xxx_NSOL_224
    new_filename=`basename ${f} |  cut -f 3-4 -d "." | sed -e  "s/hifi_kinetics_sorted_//g" `;  
    echo "${f} -> ${new_filename}"
    #echo 
    cp ${f} ../modification/${Dataset}_${strategy}/rename/${new_filename}.basemods.gff; 
done

# make motif summary into one tsv file--------------------------------------------------
Dataset=AllGenomev0.6  #*
Dataset=P-SAGv0.4      #*

strategy=v12maskNaive
strategy=v12maskMapSep
strategy=v25maskNaive
strategy=v25maskMapSep  #*

#header
cat ../modification/${Dataset}_${strategy}/original/*motifs.csv | head -1 | sed -e "s/,/\t/g" -e "s/^/GenomeID\t/g"  > ../modification/motif_${Dataset}_${strategy}.tsv

#body
for f in ../modification/${Dataset}_${strategy}/original/*motifs.csv; do   
    #v25MapSep-pbmm2HiFi_AllGenomev0.5_fixmask_SY.hifi_kinetics_sorted_P-MAGv0.4_1st_SY_013.motifs.csv  
    #v12pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_188.motifs.csv

    genomeID=`basename ${f} | rev | cut -f3-4 -d "." | rev | sed -e "s/hifi_kinetics_sorted_//g"`; #P-SAGv0.4_xxx_NSOL_188
    echo ${genomeID}
    
    #regist after add head column
    cat ${f} | grep -v "^motifString" | sed -e "s/^/${genomeID}\t/g" -e "s/,/\t/g" >> ../modification/motif_${Dataset}_${strategy}.tsv; 

done    #for output
head ../modification/motif_${Dataset}_${strategy}.tsv

# check items of which modification prediction was completed ---------------------------------------------- 
# ls ../modification/AllGenomev0.5_v25maskMapSep/original/*.check | wc
# ls ../modification/AllGenomev0.5_v25maskMapSep/original/*.basemods.gff | wc
# ls ../modification/AllGenomev0.5_v25maskMapSep/original/*.log | wc
# ls ../GenomeProkaryotes/P-AllMAGandSAGv0.5_fix/ | wc
# ls ../ReadSeparation/pbmm2HiFi_AllGenomev0.5_fixmask_TOTAL/MapSep-pbmm2HiFi_*.bam -la -sh  > ../SummaryInformation/log_mappingFileSize_AllGenomev0.5.txt
# ls ../modification/AllGenomev0.5_v25maskMapSep/original/*.check | cut -f5 -d "/" | cut -f7- -d "_" | sed -e "s/.basemods.check//g" > ../SummaryInformation/log_completedMotifsFile_AllGenomev0.5_v25maskMapSep.txt

# extract genomes which the modification analysis was completed
mkdir ../GenomeProkaryotes/P-AllMAGandSAGv0.5_modPredCompl
while read line; do
    cp ../GenomeProkaryotes/P-AllMAGandSAGv0.5_fix/${line}.fa ../GenomeProkaryotes/P-AllMAGandSAGv0.5_modPredCompl
done < ../SummaryInformation/log_completedMotifsFile_AllGenomev0.5_v25maskMapSep.txt
ls ../GenomeProkaryotes/P-AllMAGandSAGv0.5_modPredCompl | wc

# phylogeny
#./qsub_ES.sh  cpu     1  1   6 ./phylophlan3.sh ../GenomeProkaryotes/P-AllMAGandSAGv0.5_modPredCompl nucl accurate high     60   #not work
./qsub_DDBJ.sh medium 16 10 120 ./phylophlan3.sh ../GenomeProkaryotes/P-AllMAGandSAGv0.5_modPredCompl nucl accurate high     60


#--------------------------------------------------
# count modified bases on each genome
#--------------------------------------------------
output=../SummaryInformation/summary_AllGenomev0.6_baseModification.tsv; rm  ${output}; for f in ../modification/AllGenomev0.6_v25maskMapSep/rename/*.basemods.gff; do 
#output=../SummaryInformation/summary_P-SAGv0.4_baseModification.tsv;     rm  ${output}; for f in ../modification/P-SAGv0.4_v25maskMapSep/rename/*.basemods.gff;     do
    genomeID=`basename ${f} | rev | cut -f3-4 -d "." | rev | sed -e "s/hifi_kinetics_sorted_//g"`; #P-SAGv0.4_xxx_NSOL_188
    echo "Genome ID: ${genomeID}"

    #get total bases (mask)
    TotalBase=`cat ../GenomeAll/AllGenomev0.6_fixmask/${genomeID}.mask.fa | grep -v "^>" | fold -w 1 | sort | uniq -c | grep -v "N" | awk '{print $1}' | awk '{s += $1} END {print s}'`
    #TotalBase=`cat ../singlecell/P-SAGv0.4_fixmask/${genomeID}.mask.fa     | grep -v "^>" | fold -w 1 | sort | uniq -c | grep -v "N" | awk '{print $1}' | awk '{s += $1} END {print s}'`
    echo "Total base: ${TotalBase}"

    #awk -v TotalBase=${TotalBase}  '{print $1 / TotalBase}'
    cat ${f} | grep -v "^#" | cut -f1,3 | sed -e "s/_[0-9]\+\t/\t/g" | sort | uniq -c | awk -v TotalBase=${TotalBase}  '{print $2 "\t" $3 "\t" $1 "\t" TotalBase "\t" $1/TotalBase }' | grep -v "modified_base" >> ${output}
    #break
done

#--------------------------------------------------
# make motif-modified_ratio table for motif hatmap plot
# "MotifSet_XXX.tsv" file should be manually prepared.
#--------------------------------------------------
qlogin -q cpu_I -l elapstim_req=8:00:00  -X  # will increase IO access speed rather at login node
qlogin -q gpu_S -l elapstim_req=24:00:00 -X

# rm  ../motifHeatmap/work/*

#naive mapping strategy
 python3 gff_ModificationMotifPresence_table.py ../motifHeatmap/MotifSet_PMv02.tsv           ../GenomeAll/AllGenomev0.6_fix/        AllGenomev0.6_v25maskNaive     ../modification/AllGenomev0.6_v25maskNaive/rename/*.basemods.gff
#python3 gff_ModificationMotifPresence_table.py ../motifHeatmap/MotifSet_PMv02.tsv           ../singlecell/P-SAGv0.4_fix/           P-SAGv0.4_v25maskNaive         ../modification/P-SAGv0.4_v25maskNaive/rename/*.basemods.gff

#re-mapping strategy
#python3 gff_ModificationMotifPresence_table.py ../motifHeatmap/MotifSet_P-MAGv0.4_v2.tsv    ../GenomeProkaryotes/P-MAGv0.4_fix/    P-MAGv0.4_v25maskReMap         ../modification/P-MAGv0.4_v25mask/rename/*.basemods.gff
#python3 gff_ModificationMotifPresence_table.py ../motifHeatmap/MotifSet_P-MAGv0.4_v2.tsv    ../singlecell/P-SAGv0.4_fix/           P-SAGv0.4_v25maskReMap         ../modification/P-SAGv0.4_v25mask/rename/*.basemods.gff

#mapped-read sepalation strategy
 python3 gff_ModificationMotifPresence_table.py ../motifHeatmap/MotifSet_Chloroflexusv02.tsv ../singlecell/P-SAGv0.4_fix/           P-SAGv0.4_v25maskMapSep        ../modification/P-SAGv0.4_v25maskMapSep/rename/*.basemods.gff
 python3 gff_ModificationMotifPresence_table.py ../motifHeatmap/MotifSet_All_v03.tsv         ../GenomeAll/AllGenomev0.6_fix/        AllGenomev0.6_v25maskMapSep    ../modification/AllGenomev0.6_v25maskMapSep/rename/*.basemods.gff

#--------------------------------------------------
# Motifs from Chlorofrexi
cat ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_034.motifs.csv \
    ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_068.motifs.csv \
    ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_069.motifs.csv \
    ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_113.motifs.csv \
    ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_128.motifs.csv \
    ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_130.motifs.csv \
    ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_138.motifs.csv \
    ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_162.motifs.csv \
    ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_196.motifs.csv \
    ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_215.motifs.csv \
    ../modification/P-SAGv0.4_v25maskMapSep/original/v25MapSep-pbmm2HiFi_P-SAGv0.4_fixmask_NSOL.hifi_kinetics_sorted_P-SAGv0.4_xxx_NSOL_220.motifs.csv | awk '!a[$0]++' > ../modification/Chloroflexus.motifs.tsv


#===================================================================================================================================================================================================
# E.coli modification analysis
# data download: https://downloads.pacbcloud.com/public/dataset/Ecoli/egs/
# Sequel II
#===================================================================================================================================================================================================
#data download
aria2c -x10  https://downloads.pacbcloud.com/public/dataset/Ecoli/egs/m64004_200618_002500.subreads.bam 
cp ../Ecoli_PacBioOfficial/ecoli_pbi_Jan2021_majorStrain.fasta ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta

# subreads -> HiFi
qlogin -q cpu_I -l elapstim_req=8:00:00 -X
./qsub_ES.sh cpu 3 1 20 ./ccs.sh ../Ecoli_PacBioOfficial/m64004_200618_002500.subreads.bam  ../Ecoli_PacBioOfficial/ 3 0.99
rename ".3-0.99.ccs" ".hifi" *
./qsub_ES.sh cpu 1 1 6 ~/miniconda3/envs/py38/bin/ccs-kinetics-bystrandify ../Ecoli_PacBioOfficial/m64004_200618_002500.hifi.bam ../Ecoli_PacBioOfficial/m64004_200618_002500.hifi_kinetics.bam -j 60

# mapping 
./qsub_ES.sh cpu 1 1 6  ./mapping.sh -t pbmm2HiFi -b ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta -i ../Ecoli_PacBioOfficial/m64004_200618_002500.hifi_kinetics.bam -T 64 

# downsampling
samtools view -b -s 0.5   ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.bam -o ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.050.bam  -@ 10
samtools view -b -s 0.1   ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.bam -o ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.010.bam  -@ 10
samtools view -b -s 0.05  ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.bam -o ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.005.bam  -@ 10
samtools view -b -s 0.01  ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.bam -o ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.001.bam  -@ 10
samtools view -b -s 0.005 ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.bam -o ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.0005.bam -@ 10
samtools view -b -s 0.001 ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.bam -o ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.0001.bam -@ 10

#modification detection
./qsub_ES.sh     cpu   3 1  24 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.bam      60 v25 # not complete until the deadline
./qsub_ES.sh     cpu   3 1  24 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.050.bam  60 v25 # not complete until the deadline
./qsub_ES.sh     cpu   3 1  24 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.010.bam  60 v25 # not complete until the deadline
./qsub_ES.sh     cpu   3 1  24 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.005.bam  60 v25 # not complete until the deadline
./qsub_ES.sh     cpu   3 1  24 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.001.bam  60 v25
./qsub_ES.sh     cpu   3 1  24 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.0005.bam 60 v25 # 1h
./qsub_ES.sh     cpu   3 1  24 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.0001.bam 60 v25 # 0.5h

./sbatch_DDBJ.sh rome 10 8 360 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.bam      16 v25
./sbatch_DDBJ.sh rome 10 8 150 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.050.bam  16 v25
./sbatch_DDBJ.sh rome 10 8 120 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.010.bam  16 v25
./sbatch_DDBJ.sh rome 10 8 72  ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.005.bam  16 v25
./sbatch_DDBJ.sh rome 10 8 36  ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.001.bam  16 v25
#./sbatch_DDBJ.sh rome 10 8 36 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.0005.bam 10 v25
#./sbatch_DDBJ.sh rome 10 8 36 ./modification_pacbio.sh ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta ../mapping/EcoliK12MG1655/pbmm2HiFi_EcoliK12MG1655_m64004_200618_002500.hifi_kinetics_sorted.0001.bam 10 v25

mkdir ../modification/EcoliK12MG1655
mv    ../modification/v25pbmm2HiFi_EcoliK12MG1655* ../modification/EcoliK12MG1655

# count modified bases on each genome
output=../SummaryInformation/summary_Ecoli_baseModification.tsv; rm  ${output}; for f in ../modification/EcoliK12MG1655/*.basemods.gff; do 
output=../SummaryInformation/summary_Ecoli_baseModification.tsv; rm  ${output}; for f in ../modification/*.basemods.gff; do 
    genomeID=`basename ${f} | rev | cut -f3-4 -d "." | rev | sed -e "s/hifi_kinetics_sorted_//g"`; #P-SAGv0.4_xxx_NSOL_188
    echo "Genome ID: ${genomeID}"

    #get total bases (mask)
    TotalBase=`cat ../Ecoli_PacBioOfficial/EcoliK12MG1655.fasta | grep -v "^>" | fold -w 1 | sort | uniq -c | grep -v "N" | awk '{print $1}' | awk '{s += $1} END {print s}'`
    echo "Total base: ${TotalBase}"

    cat ${f} | grep -v "^#" | cut -f1,3 | sed -e "s/_[0-9]\+\t/\t/g" | sort | uniq -c | awk -v TotalBase=${TotalBase}  '{print $2 "\t" $3 "\t" $1 "\t" TotalBase "\t" $1/TotalBase }' | grep -v "modified_base"  >> ${output}
    #break
done
cat ${output}

#===================================================================================================================================================================================================
# MTase enzymatic analysis using one-pot (in vitro) or pCold + E. coli (in vivo) methods
#===================================================================================================================================================================================================
seqkit split ../assay/20241018_1st_all.faa -s 1

# DiANNA (not work)
conda activate py27
python2 ~/software/DiANNAstandalone/ModuleABC_standalone.py ../assay/20241018_1st_all.faa.split/20241018_1st_all.part_001.faa 1
python2 ~/software/DiANNAstandalone/ModuleABC_standalone.py ../assay/20241018_1st_all.faa 2
for f in ../assay/20241018_1st_all.faa.split/20241018_1st_all.part_*.faa ; do
    python2 ~/software/DiANNAstandalone/ModuleABC_standalone.py ${f} 1
done

# manual check for specific motifs
cat ../GenomeProkaryotes/P-MAGv0.4_fix/P-MAGv0.4_1st_NSBR_Circle_009.fa | seqkit locate -idp CAGNNNNNNTAYC | cut -f7 | sort | uniq -c | sort -rn
cat ../GenomeProkaryotes/P-MAGv0.4_fix/P-MAGv0.4_1st_NSGRY_004.fa       | seqkit locate -idp GCHGC         | cut -f7 | sort | uniq -c | sort -rn

# count motifs on the constracts
cat Case01_pColdIII.fasta | seqkit locate -idp AGCT     | grep -v seqID | wc
cat Case03_pColdIII.fasta | seqkit locate -idp CCGG     | grep -v seqID | wc

cat Case04_pColdIII.fasta | seqkit locate -idp CGCCTG   | grep -v seqID | wc
cat Case04_pColdIII.fasta | seqkit locate -idp GGCGCC   | grep -v seqID | wc
cat Case04_pColdIII.fasta | seqkit locate -idp CCTGCAGG | grep -v seqID | wc

cat Case06_pColdIII.fasta | seqkit locate -idp CTCGAG   | grep -v seqID | wc
cat Case07_pColdIII.fasta | seqkit locate -idp CTCGAG   | grep -v seqID | wc
cat Case12_pColdIII.fasta | seqkit locate -idp GCHGC    | grep -v seqID | wc

cat Case15_pColdIII.fasta | seqkit locate -idp TCGA     | grep -v seqID | wc
cat Case15_pColdIII.fasta | seqkit locate -idp GTCGAC   | grep -v seqID | wc

cat Case16_pColdIII.fasta | seqkit locate -idp CTGCAG   | grep -v seqID | wc
cat Case16_pColdIII.fasta | seqkit locate -idp ATGCAT   | grep -v seqID | wc
cat Case16_pColdIII.fasta | seqkit locate -idp GTGCAG   | grep -v seqID | wc

cat Case17_pColdIII.fasta | seqkit locate -idp CTGCAG   | grep -v seqID | wc
cat Case17_pColdIII.fasta | seqkit locate -idp ATGCAT   | grep -v seqID | wc
cat Case17_pColdIII.fasta | seqkit locate -idp GTGCAG   | grep -v seqID | wc

cat Case18_pColdIII.fasta | seqkit locate -idp CTGCAG   | grep -v seqID | wc
cat Case18_pColdIII.fasta | seqkit locate -idp ATGCAT   | grep -v seqID | wc
cat Case18_pColdIII.fasta | seqkit locate -idp GTGCAG   | grep -v seqID | wc

#signalP --------------------------------------
./signalp.sh ../assay/GeneSetHotSpring1.faa      10
./signalp.sh ../assay/GeneSetHiraokaNAR2022.faa  10

#AF2--------------------------------------
mkdir  ../alphafold/{query_origin,query_base} -p
rm     ../alphafold/query_origin/*
rm     ../alphafold/query_base/*.faa
seqkit split -s 1 ../assay/GeneSet1/20241018_1st_all.faa          -O ../alphafold/query_origin 
seqkit split -s 1 ../assay/MTaseAssay_MarineMetaepigenome_fix.faa -O ../alphafold/query_origin 

#rename filename and seqname
for f in ../alphafold/query_origin/*.faa; do
#    name=`head -1 ${f} | cut -c2- | cut -f1 -d " "`
    name=`head -1 ${f} | cut -c2- | cut -f1 -d " " | sed -e "s/0.4/04/g" -e "s/|/_/g"`
    echo "${f} -> ${name}"
    echo ">${name}"  >  ../alphafold/query_base/${name}.faa
    tail -n +2 ${f}  >> ../alphafold/query_base/${name}.faa
done

#alphafolds
for f in ../alphafold/query_base/*.faa ;  do 
    BASE_FILENAME=` basename ${f} | sed -e "s/.faa//g" `
    if [ -e ../alphafold/work/${BASE_FILENAME}/ranked_0.pdb ]; then
        echo "Structure file already exist. Skip. ${f}"
        continue 
    fi
    ./qsub_ES.sh gpu 1 1 6  ./alphafold2.sh ${f}; 
done

#clean
for f in ../alphafold/work/*; do
    rm -r ${f}/msas
done

#copy
mkdir ../alphafold/summary_pdb/ -p
for f in ../alphafold/work/*; do
    FILENAME=${f##*/}
    echo ${f}
    cp ${f}/ranked_0.pdb  ../alphafold/summary_pdb/${FILENAME}.pdb -n
done


#===================================================================================================================================================================================================
# Public Viral genome analysis
#===================================================================================================================================================================================================
# download

# gene call

# MTase prediction




